<!DOCTYPE html>

<html><head>
    <meta http-equiv="Content-type" content="text/html; charset=UTF-8">
    <title>Kelvin - Mathematical Medicine LLC</title>
    <link rel="stylesheet" href="https://mathmed.org/mm.css" type="text/css">
    <link rel="apple-touch-icon" sizes="180x180" href="https://mathmed.org/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="https://mathmed.org/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="https://mathmed.org/favicon-16x16.png">
    <link rel="manifest" href="/site.webmanifest">
</head><body>
    <div class="tabs"><div class="tab" style="display: block">
        <h1><span class="sc">Kelvin</span></h1>
        <h1 style="float: right; padding-left: 0.5em; width: 20%"><a href="https://mathmed.org"><img alt="Mathematical Medicine LLC" src="LogoAndWordmark-Vertical.png"></a></h1>
        <p><span class="sc">Kelvin</span> is a flexible program for measuring statistical evidence based on genetic data. It is based on the PPL framework and produces output on the posterior probability (0,..,1) scale. Primary outcome statistics include the PPL (posterior probability of linkage (L)), as well as various forms of the PPLD (posterior probability of linkage disequilibrium (LD)). In the context of analysis between a trait locus and a marker, the PPLD models allelic association due to very close genomic proximity. <span class="sc">Kelvin</span> handles both dichotomous and quantitative trait data, and can accept as input multigenerational pedigree data, nuclear families, “trios,” and case-control data. The program supports parent-of-origin effects, co-variate dependent penetrances (or means), and two-locus epistasis. <span class="sc">Kelvin</span> also includes special routines for sequentially updating results in order to accumulate evidence across multiple sets of data.</p>
        
        
        <h3>Documentation and References</h3>
        
        <p><a href="doc/">Documentation for Kelvin is available here.</a></p>
        
        <p>References:</p>
        <ul>
            <li>Seok S-C, Evans M, Vieland VJ. <a href="https://doi.org/10.1089/cmb.2008.0175">Fast and accurate calculation of a computationally intensive statistic for mapping disease genes.</a> J Comput Biol, 16(5):659-676, 2009. PMID: 19432537, PMC3148122, doi:10.1089/cmb.2008.0175</li>
            <li>Vieland VJ, Huang Y, Seok S-C, Burian J, Catalyurek U, O’Connell J, Segre A, Valentine-Cooper W. <a href="https://doi.org/10.1159/000330634">KELVIN: a software package for rigorous measurement of statistical evidence in human genetics.</a> Hum Hered 72(4):276-288, 2011. PMID: 22189470, PMC3267994, doi:10.1159/000330634</li>
            <li>Huang Y, Thomas A, Vieland VJ. <a href="https://doi.org/10.3389/fgene.2013.00059">Employing MCMC under the PPL framework to analyze sequence data in large pedigrees.</a> Frontiers in Applied Genetic Epidemiology 4:59, 2013. PMID: 23626600, PMC363090, doi:10.3389/fgene.2013.00059</li>
        </ul>
        
        <h3>Where To Get It</h3>
        
        <p>The latest release is <a href="https://github.com/MathematicalMedicine/kelvin/releases/latest">available on Github</a>. You can also <a href="https://github.com/MathematicalMedicine/kelvin/">browse our repository</a>.
    </div>    
</body></html>
