<html>
<head>
  <title>Kelvin Configuration Directives</title>
  <style type='text/css'>
    div.hang {padding-left: 2em ; text-indent: -2em }
  </style>
</head>
<body>
<h1>Kelvin Configuration Directives</h1>

If you are just getting started with kelvin, or are uncomfortable directly editing a
configuration file, you can use the <a href="http://hodgkin.ccri.net/software/gKelvin/">graphic
configuration file editor</a> instead.

<h2>The Configuration File</h2>

Kelvin's configuration file format is intended to be easy to read and edit. Kelvin, in
general, requires the user to explicitly specify what Kelvin is to do, and doesn't make
any assumptions about what the user may or may not want. Kelvin will display helpful error
messages and terminate if the configuration file contains incompatible directives. 

<h3>Configuration File Format</h3>
The kelvin configuration file is a text file containing directives that describe the nature
of the analysis to be performed, and the names of input and output files. Directives typically
appear on separate lines, but multiple directives, separated by a semicolon (';'), may appear
on the same line. Some directives require arguments, which must appear on the same line as
the directive they modify. Blank lines in the configuration file are ignored. Extra white
space (spaces and tabs) around directives and arguments is ignored. Comments may be 
embedded in a configuration file with a pound sign ('#'). The pound sign and any text, up
to the end of the line, will be ignored. Comments may appear on a line by themselves,
or at the end of a line that also contains a directive(s). In the latter case, a semicolon
is not required to separate the directive and the comment.

<h3>Directives</h3>
Directives are words or combinations of words with no intervening spaces. This document
will capitalize the individual words that make up a directive (for example, <tt>MapFile</tt>),
but they may appear in the configuration file with any capitalization (<tt>Mapfile</tt>,
<tt>mapfile</tt>, <tt>MAPFILE</tt>, etc.). Also, directives may be abbreviated, so long
as the abbreviation is unique. For example, <tt>MapFile</tt> could be abbreviated to
<tt>Map</tt>. <tt>SexLinked</tt> could be abbreviated to <tt>SexL</tt>, but not to
<tt>Sex</tt>, since that could also be an abbreviation for <tt>SexSpecific</tt>. 

<h3 id='arguments'>Arguments</h3>
Most directives require one or more arguments, which will take one of the following forms:
<ul>
  <li>A Filename - a character string that must be a valid filename, optionally
    including a relative or absolute path preceding the filename. Generally, Kelvin
    will abort if a file given as input does not exist, and will overwrite a file
    given for output if it does exist. The total length of a filename may not
    exceed 256 characters.</li>
  <li>A Literal String - a few directives require a specific string as an argument.
    These strings may appear with any capitalization, but they should not be abbreviated.</li>
  <li>A Value - an integer or real number, depending on the directive.</li>
  <li id='range'>A Range - a shorthand method for specifiying a series of values with a
    regular interval. A range is specified as a start value, a literal '<tt>-</tt>',
    an end value, a literal '<tt>:</tt>', and an increment. Start, end and increment values
    may be integers or real numbers. Start and end values may be positive or negative, with
    the restriction that the end must be no less than the start. The increment must be
    a positive value.  Intervening whitespace may be added for clarity, but is not necessary.
    Kelvin will expand the range into a list of values that begins with the start value,
    increases by the increment, and ends with the last value that is no greater than the
    end value. For example, a range that runs from -1 to 1 in increments of 0.1 would be
    specified as '<tt>-1 - 1 : 0.1</tt>'. A range specified as '<tt>0-1:0.15</tt>' would
    expand to <tt>0, 0.15 ... 0.75, 0.90</tt>.</li>
</ul>


<h3>Sample Configuration File</h3>
Kelvin configuration files can be quite simple. Here's a file for a multipoint 
analysis that will generate PPLs for a dichotomous trait at one centiMorgan
intervals, considering 4 markers around each trait position:

<table cellspacing=2>
  <tr><td align='right' style='background: #888888'><tt> 1 </tt></td>
    <td><tt># Sample multipoint, dictomous trait configuration</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 2 </tt></td>
    <td></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 3 </tt></td>
    <td><tt>MultiPoint 4</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 4 </tt></td>
    <td><tt>TraitPositions 0-end:1</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 5 </tt></td>
    <td><tt></tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 6 </tt></td>
    <td><tt>PhenoCodes 0, 1, 2</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 7 </tt></td>
    <td><tt></tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 8 </tt></td>
    <td><tt>PedigreeFile ped.post</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 9 </tt></td>
    <td><tt>LocusFile locus.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 10 </tt></td>
    <td><tt>MapFile map.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 11 </tt></td>
    <td><tt>FrequencyFile freq.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 12 </tt></td>
    <td><tt>BayesRatioFile br.out</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 13 </tt></td>
    <td><tt>PPLFile ppl.out</tt></td></tr>
</table>

Lines 3 and 4 specify a 4-marker multipoint analysis, and that the trait
should be placed at 1 cM intervals, starting at 0 cM and running through
the end of the map. Line 6 specifies the values that will appear in the
affection status column in the pedigree file. Lines 8-13 specify
input and output files.<p>

Here's a file for two-point analysis that will generate PPLs and
linkage disequilibrium statistics between a dichotomous trait and each
marker in the dataset:

<table cellspacing=2>
  <tr><td align='right' style='background: #888888'><tt> 1 </tt></td>
    <td><tt># Sample two-point, dictomous trait LD configuration</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 2 </tt></td>
    <td></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 3 </tt></td>
    <td><tt>LD</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 4 </tt></td>
    <td><tt></tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 5 </tt></td>
    <td><tt>PhenoCodes 0, 1, 2</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 6 </tt></td>
    <td><tt></tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 7 </tt></td>
    <td><tt>PedigreeFile ped.post</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 8 </tt></td>
    <td><tt>LocusFile locus.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 9 </tt></td>
    <td><tt>MapFile map.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 10 </tt></td>
    <td><tt>FrequencyFile freq.dat</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 11 </tt></td>
    <td><tt>BayesRatioFile br.out</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 12 </tt></td>
    <td><tt>PPLFile ppl.out</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 13 </tt></td>
    <td><tt>MODFile mod.out</tt></td></tr>
  <tr><td align='right' style='background: #888888'><tt> 14 </tt></td>
    <td><tt>ExtraMODs</tt></td></tr>
</table>

Two-point analysis is the default, so no directive is necessary to specify
that. Line 3 specifies a linkage disequilibrium analysis. Line 5 specifies
the values that will appear in the affection status field of the pedigree.
Lines 7-13 specify input and output files. Line 14 specifies that additional
information should be written to the MOD file specified on line 13.<p>

<h2 id="analysis">Types of Analysis</h2>

Kelvin supports a wide variety of types of analyses. These types can be broken
into general categories, with a small number of options for each category.
Some options are compatible with other options, some are not.

<h3 id='twopoint_vs_multipoint'>Two-Point vs. Multipoint</h3>
Two-point analysis is the default. Multipoint analysis is enabled with the
<a href='#multipoint'>Multipoint</a> directive. Multipoint analysis is
incompatible with <a href='#ld_vs_le>linkage disequilibrium</a> and 
<a href='#traitmarker_vs_markermarker>marker-to-marker</a> analyses.

<h3 id='ld_vs_le'>Linkage Disequilibrium vs. Linkage Equilibrium</h3>
Linkage equilibrium is the default for <a href='#twopoint_vs_multipoint'>
two-point</a> analyses, and is the only option for
<a href='#twopoint_vs_multipoint'>multipoint</a> analyses. Linkage
Disequilibrium can be enabled for two-point analyses with the
<a href='#ld'>LD</a> directive.

<h3 id='dt_vs_qt_vs_qtt'>Dichotomous Trait vs. Quantitative Trait vs. Quantitative Trait With Threshold</h3>
A dichotomous trait model is the default. A quantitative trait model can
be specified with the <a href='#qt'>QT</a> directive. A quantitative trait
model with a threshold can be specified with the <a href='#qtt'>QTT</a>
and <a href='#threshold'>Threshold</a> directives. Quantitative
trait directives are incompatible with, and the default dichotomous trait
model is disabled for, <a href='#traitmarker_vs_markermarker'>
marker-to-marker</a>.

<h3 id='sexave_vs_sexspec'>Sex Averaged vs. Sex Specific</h3>
By default, Kelvin will perform its calculations using the sex-averaged
centiMorgan marker positions in the map file. If sex-specific marker positions
are available, Kelvin can be made to use those with the <a href='#sexspecific'>
SexSpecific</a> directive. Sex-specific maps are not supported for
<a href='#ld_vs_le'>LD</a> analyses.

<h3 id='imprinting_vs_not'>Imprinting vs. Non-imprinting</h3>
Kelvin will ignore the possibility of imprinting (parent-of-origin)
effects by default. Imprinting effects can be considered by specifying the
<a href=#'imprinting'>Imprinting</a> directive. Imprinting analysis is
incompatible with <a href='#traitmarker_vs_markermarker'>marker-to-marker</a>
analyses.

<h3 id='traitmarker_vs_markermarker'>Trait-to-Marker vs. Marker-to-Marker</h3>
Trait-to-marker analysis is the default, and considers the relationship
between a hypothetical trait locus and a marker or group of markers.
Marker-to-marker analysis is enabled with the <a href='#markertomarker'>
MarkerToMarker</a> directive, and considers the relationship between
pairs of markers only. Since no trait is considered at all under marker-to-marker,
<a href='#dt_vs_qt_vs_qtt'>QT, QTT</a> and <a href='#imprinting_vs_not'>
imprinting</a> are incompatible. <a href='#twopoint_vs_multipoint'>Multipoint</a>
analyses are also incompatible with marker-to-marker.

<p>

<h2 id="summary">Directive Summary</h2>
<table cellspacing=5>
<tr>
  <td><b>Input And Output</b></td>
  <td><b>Analysis Type</b></td>
  <td><b>Trait Model</b></td>
  <td><b>Options</b></td>
</tr>
<tr>
  <td valign='top'>
    <a href='#pedigreefile'>PedigreeFile</a><br>
    <a href='#locusfile'>LocusFile</a><br>
    <a href='#frequencyfile'>FrequencyFile</a><br>
    <a href='#mapfile'>MapFile</a><br>
    <a href='#bayesratiofile'>BayesRatioFile</a><br>
    <a href='#pplfile'>PPLFile</a><br>
    <a href='#modfile'>MODFile</a><br>
    <a href='#extramods'>ExtraMODs</a>
  </td>
  <td valign='top'>
    <a href='#multipoint'>MultiPoint</a><br>
    <a href='#ld'>LD</a><br>
    <a href='#markertomarker'>MarkerToMarker</a><br>
  </td>
  <td valign='top'>
    <a href='#qt'>QT</a><br>
    <a href='#qtt'>QTT</a><br>
    <a href='#degreesoffreedom'>DegreesOfFreedom</a><br>
    <a href='#threshold'>Threshold</a><br>
    <a href='#liabilityclasses'>LiabilityClassses</a><br>
  </td>
  <td valign='top'>
    <a href='#phenocodes'>PhenoCodes</a><br>
    <a href='#sexlinked'>SexLinked</a><br>
    <a href='#sexspecific'>SexSpecific</a><br>
    <a href='#imprinting'>Imprinting</a><br>
    <a href='#traitpositions'>TraitPositions</a><br>
    <a href='#diseasealleles'>DiseaseAlleles</a><br>
    <a href='#polynomialscale'>PolynomialScale</a><br>
    <a href='#nonpolynomial'>NonPolynomial</a><br>
    <a href='#dryrun'>DryRun</a><br>
    <a href='#log'>Log</a><br>
  </td>
</tr>
</table>

<h2 id='reference'>Directive Reference</h2>

In this reference, the following convetions are used to describe the valid arguments 
to directives.

<ul>
  <li>Arguments that appear in <tt>roman type</tt> are <i>literals</i>. If they
    are to be used in the configuration file, they must appear as written in this
    reference. For example,<p>
    <tt>QT ChiSq</tt><p>
    indicates that the literal string <tt>ChiSq</tt> is a legal argument to the
    <tt>QT</tt> directive.
  </li>
  <li>Arguments that appear in <tt><i>italic type</i></tt> are <i>symbolic</i>, and
    refer to one of the four types of <a href='#arguments'>arguments</a>, described
    above. If they are to be used in the configuration file, they must conform to the
    description of the respective argument type. For example,<p>
    <tt>PedigreeFile <i>filename</i></tt><p>
    indicates that the <tt>PedigreeFile</tt> directive requires a single argument
    that conforms to the <a href='#arguments'>filename</a> type.
  </li>
  <li>A group of arguments inside brackets, separated by vertical bars, indicates a
    choice between multiple possible arguments, of which exactly one must be chosen.
    The individual options will either be literal or symbolic. For example,<p>
    <tt>MarkerToMarker [ All | Adjacent ]</tt><p>
    indicates that the <tt>MarkerToMarker</tt> directive can take either of the
    literals <tt>All</tt> or <tt>Adjacent</tt> as arguments.
  </li>
  <li>Arguments that appear inside braces are optional, but will otherwise be
    literal, symbolic or a group of arguments, as described above. Braces containing
    an ellipsis means that whatever came immediately before the ellipsis may be 
    optionally repeated one or more times. For example,<p>
    <tt>QTT Normal { <i>mean</i>, <i>standarddeviation</i> }</tt><p>
    indicates that the <tt>QTT</tt> directive with the <tt>Normal</tt> argument
    optionally accepts two additional arguments, <tt><i>mean</i></tt> and
    <tt><i>standarddeviation</i></tt>. In this example,<p>
    <tt>TraitPositions [ Marker | <i>value</i> | <i>range</i> ] {, ... }</tt><p>
    the <tt>TraitPositions</tt> directive can take either a literal <tt>Marker</tt>,
    a single <tt><i>value</i></tt> or a <tt><i>range</i></tt>, and furthermore
    can take one or more of any of the preceding, after a separating comma.
  </li>
</ul>

    <div class='hang' id='bayesratiofile'>
      <b>BayesRatioFile</b><br>
      <tt>BayesRatioFile <i>filename</i></tt><br>
      Specifies the name of the output file to which Bayes ratios will be written.
      For <a href='#twopoint_vs_multipoint'>multipoint</a> analyses, PPLs will also
      be written to this file. For <a href='#traitmarker_vs_markermarker'>
      Marker-to-marker</a> analyses, no output is written to this file. Defaults
      to <tt>br.out</tt> if not specified.
    </div><p>

    <div class='hang' id='degreesoffreedom'>
      <b>DegreesOfFreedom</b><br>
      <tt>DegreesOfFreedom <i>min</i>, <i>max</i></tt><br>
      Specifies the minimum and maximum degrees of freedom when using the Chi-squared
      distribution for quantitative trait analysis, if needed. The default is for the
      minimum and maximum degrees of freedom to be set using the smallest and largest
      values from the trait data in the pedigree file. If those values are less then 0
      or greater than 30, then this directive must be specified. The <tt><i>min</i></tt>
      and <tt><i>max</i></tt> should be specified as <a href='#arguments'>values</a>. This
      directive is only legal if one of the <a href='qt'>QT</a> or <a href='qtt'>QTT</a>
      directives is specified with the Chi-squared distribution.
    </div><p>

    <div class='hang' id='diseasealleles'>
      <b>DiseaseAlleles</b><br>
      <tt>DiseaseAlleles <i>number</i></tt><br>
      Specifies the number of alleles for the hypothetical trait locus. Currently
      the only supported number of disease alleles is 2, which also the default.
      This directive is incompatible with the <a href='#markertomarker'>MarkerToMarker</a>
      directive.
    </div><p>

    <div class='hang' id='dryrun'>
      <b>DryRun</b><br>
      <tt>DryRun</tt><br>
      Specifies that configuration and input data should be read and validated, and
      that the complexity of the analysis should be estimated, but no calculations
      are performed.
    </div><p>

    <div class='hang' id='extramods'>
      <b>ExtraMODs</b><br>
      <tt>ExtraMODs</tt><br>
      Specifes that additional data should be written to the file specified by the
      <a href='#modfile'>MODFile</a> directive. For <a href='#twopoint_vs_multipoint'>
      two-point</a> analyses, an extra line containing MOD information where Theta is 0
      will be written, and for <a href='#ld_vs_le'>LD</a> analyses, another line 
      containing MOD information where D' is 0 will be written. This directive is
      incompatible with the <a href='#Multipoint'>Multipoint</a> directive.
    </div><p>

    <div class='hang' id='frequencyfile'>
      <b>FrequencyFile</b><br>
      <tt>FrequencyFile <i>filename</i></tt><br>
      Specifies the name of the input file from which marker allele frequencies
      will be read. Defaults to <tt>markers.dat</tt> if not specified.
    </div><p>

    <div class='hang' id='imprinting'>
      <b>Imprinting</b><br>
      <tt>Imprinting</tt><br>
      Specifies that imprinting, or parent-of-origin, effects on the trait should
      be considered. This directive is incompatiable with the 
      <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='ld'>
      <b>LD</b><br>
      <tt>LD</tt><br>
      Specifies that linkage disequilibrium statistics should be calculated. This
      directive is incompatible with the <a href='#multipoint'>MultiPoint</a>
      and <a href='#sexspecific'>SexSpecific</a> directives.
    </div><p>

    <div class='hang' id='liabilityclasses'>
      <b>LiabilityClassses</b><br>
      <tt>LiabilityClasses <i>number</i></tt><br>
      Specifies that individuals in the pedigrees have been categorized into two
      or more classes. The penetrance values for individuals in one class will
      be considered independently from the other class(es). Note that a column that
      identifies the class for each individual must appear in the pedigree file.
      Kelvin expects liability classes to be numbered contiguously, starting at 1.
      If Kelvin detects that one or more of the specified classes is empty, it will
      drop the empty classes, and consolidate the remaining classes. This directive
      is incompatiable with the <a href='#markertomarker'>MarkerToMarker</a>
      directive.
    </div><p>

    <div class='hang' id='locusfile'>
      <b>LocusFile</b><br>
      <tt>LocusFile <i>filename</i></tt><br>
      Specifies the name of the input file from which the pedigree column 
      layout information will be read. This file is sometimes referred to as a 
      'data file.' Defaults to <tt>datafile.dat</tt> if not specified.
    </div><p>

    <div class='hang' id='log'>
      <b>Log</b><br>
      <tt>Log <i>facility</i> <i>level</i></tt><br>
      Specifies that Kelvin should generate extra output while it runs. This is 
      generally useful only for debugging purposes. Under the default settings,
      Kelvin is adequately verbose to inform the user of any configuration or
      data errors, and for the user to keep track of progress.<p>
    </div><p>

    <div class='hang' id='mapfile'>
      <b>MapFile</b><br>
      <tt>MapFile <i>filename</i></tt><br>
      Specifies the name of the input file from which the marker map data
      will be read. Defaults to <tt>mapfile.dat</tt> if not specified.
    </div><p>

    <div class='hang' id='markertomarker'>
      <b>MarkerToMarker</b><br>
      <tt>MarkerToMarker [ All | Adjacent ]</tt><br>
      Specifies that only the relationships of pairs of markers should be considered,
      and that no trait is included in the analisys. Under linkage equilibrium,
      the intermarker distance is estimated. Under <a href='#ld'>linkage disequilibrium</a>,
      the degree of linkage disequilibrium between the markers is estimated.
      If <tt>All</tt> is specified, all possible pairs of markers in the dataset
      are considered. If <tt>Adjacent</tt> is specified, only adjacent pairs of
      markers are considered.
    </div><p>

    <div class='hang' id='modfile'>
      <b>MODFile</b><br>
      <tt>MODFile <i>filename</i></tt><br>
      Specifies the name of the output file to which maximizing model
      information will be written. This directive is incompatible with the 
      <a href='#multipoint'>MultiPoint</a> directive.
    </div><p>

    <div class='hang' id='multipoint'>
      <b>MultiPoint</b><br>
      <tt>MultiPoint <i>count</i></tt><br>
      Specifies that Kelvin should perform a multipoint analysis, generating a
      multipoint PPL for each location specified with the <a href='#traitpositions'>
      TraitPositions</a> directive. Groups of <tt><i>count</i></tt> adjacent markers,
      centered on each trait location, will be used to calculate the PPL. This directive
      is incompatible with the <a href='#markertomarker'>MarkerToMarker</a> and
      <a href='#ld'>LD</a> directives.
    </div><p>

    <div class='hang' id='nonpolynomial'>
      <b>NonPolynomial</b><br>
      <tt>NonPolynomial</tt><br>
      Specifies that Kelvin should not internally represent the likelihood
      caclulation as a polynomial. This option will dramatically reduce memory
      requirements, but can increase run time by orders of magnitude for 
      complex analyses. This directive is incompatible with the
      <a href='#polynomialscale'>PolynomialScale</a> directive.
    </div><p>

    <div class='hang' id='pedigreefile'>
      <b>PedigreeFile</b><br>
      <tt>PedigreeFile <i>filename</i></tt><br>
      Specifies the name of the input file from which the pedigree data
      will be read. Defaults to <tt>pedfile.dat</tt> if not specified.
    </div><p>

    <div class='hang' id='phenocodes'>
      <b>PhenoCodes</b><br>
      <tt>PhenoCodes <i>unknown</i></tt><br>
      <tt>PhenoCodes <i>unknown</i>, <i>unaffected</i>, <i>affected</i></tt><br>
      Specifies the values that will appear in the affection status (phenotype) field
      of the pedigree file, to indicate if an individual is affected, unaffected, or
      if the affection status is unknown. For quantitative trait analyses, only the value
      for unknown affection status should be specified. For dichotomous 
      traits, the defaults are 0 (for unknown), 1 (for unaffected) and 2 (for affected).
      For quantitative threshold traits, the defaults are -99.99, -88.88 and 88.88.
      For quantitative traits, the default is -99.99 (for unknown). This directive
      is incompatible with the <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='polynomialscale'>
      <b>PolynomialScale</b><br>
      <tt>PolynomialScale <i>scale</i></tt><br>
      If the user knows in advance that an analysis will require large amounts of
      memory, this directive can be specified, and Kelvin will pre-allocate memory
      before it begins calculations. This will result in somewhat more efficient
      use of memory, compared to Kelvin incrementally allocating memory as it runs.
      The <tt><i>scale</i></tt> is an integer between 1 and 10. This directive is 
      incompatible with the <a href='#nonpolynomial'>NonPolynomial</a> directive.
    </div><p>

    <div class='hang' id='pplfile'>
      <b>PPLFile</b><br>
      <tt>PPLFile <i>filename</i></tt><br>
      For two-point analyses, specifies the name of the output file to which PPLs
      will be written. For multipoint analyses, PPLs are written to the file specified
      by the <a href='#bayesratiofile'>BayesRatioFile</a> directive, and no data will
      written to the file specified with this directive. Defaults to <tt>ppl.out</tt>
      if not specified.
    </div><p>

    <div class='hang' id='qt'>
      <b>QT</b><br>
      <tt>QT Normal { <i>mean</i>, <i>standarddeviation</i> }</tt><br>
      <tt>QT ChiSq</tt><br>
      Specifies that the trait is quantitative, rather than dichotomous. The trait data
      in the pedigree file should be of a quantitative nature (values on a continuous
      scale, rather than discrete affected or unaffected values). Two distributions are
      available, Normal and Chi-squared. If <tt>Normal</tt> is specified, the
      <tt><i>mean</i></tt> and <tt><i>standarddeviation</i></tt> <a href='#arguments'>values</a>
      may be provided. Otherwise, a sample mean and standard deviation will be
      calculated using the trait values in the pedigree file. If <tt>ChiSq</tt> is specified,
      the <a href='#degreesoffreedom'>DegreesOfFreedom</a> directive may be specified.
      Otherwise, the minimum and maximum degrees of freedom will be set using the
      smallest and largest values from trait data in the pedigree file. If those values
      are less than 0 or gerater than 30, then the minimum and maximum degrees of freedom
      must be set with the <a href='#degreesoffreedom'>DegreesOfFreedom</a> directive.
      This directive is incompatible with the <a href='#markertomarker'>MarkerToMarker</a>
      directive.
    </div><p>

    <div class='hang' id='qtt'>
      <b>QTT</b><br>
      <tt>QTT Normal { <i>mean</i>, <i>standarddeviation</i> }</tt><br>
      <tt>QTT ChiSq</tt><br>
      Like the <a href='#qt'>QT</a> directive, but in addition considers a threshold beyond
      which an individual as simply assumed to be affected. The minimum and maximum
      threshold may specified with the <a href='#threshold'>Threshold</a> directive. 
      Otherwise, the minimum and maximum threshold will be set using the smallest and
      largest values from the trait data in the pedigree file. If those values are less than
      0 or greater than 30, then the minimum and maximum threshold must be set with the 
      <a href='#threshold'>Threshold</a> directive. This directive in incompatible with 
      the <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='sexlinked'>
      <b>SexLinked</b><br>
      <tt>SexLinked</tt><br>
      Specifies that the pedigree contains data for the X chromosome.
    </div><p>

    <div class='hang' id='sexspecific'>
      <b>SexSpecific</b><br>
      <tt>SexSpecific</tt><br>
      Specifies that the <a href='#mapfile'>map file</a> contains sex-specific
      centiMorgan positions for the markers, and that the sex-specific positions
      should be used instead of sex-averaged positions during the analysis. Requires
      that the map file actually contains sex-specific positions. This directive is
      incompatible with the <a href='#ld'>LD</a> directive.
    </div><p>

    <div class='hang' id='threshold'>
      <b>Threshold</b><br>
      <tt>Threshold <i>min</i>, <i>max</i></tt><br>
      Specifies the minimum and maximum <a href='#arguments'>values</a> for the threshold
      for use with the  <a href='#qtt'>QTT</a> directive, if needed. The default is for
      the minimum and maximum threshold values to be set using the smallest and largest
      values from the trait data in the pedigree file. If those values are less than 0 
      or greater than 30, then this directive must be specified. The <tt><i>min</i></tt>
      and <tt><i>max</i></tt> should be <a href='arguments'>values</a>. This directive
      requires the <a href='#qtt'>QTT</a> directive.
    </div><p>

    <div class='hang' id='traitpositions'>
      <b>TraitPositions</b><br>
      <tt>TraitPositions [ Marker | <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the positions at which PPLs will be calculated in a multipoint analysis.
      The literal string <tt>Marker</tt> indicates that PPLs should be calculated at
      each marker position, in addition to any other positions. If a <a href='#range'>range</a>
      is specified, the end value of the range may be specified as the literal <tt>end</tt>.
      In that case, Kelvin will expand the range so the last value is just past the last
      marker in the map. The literal <tt>end</tt> in a range specification is only valid
      with this directive. This directive requires the <a href='#multipoint'>MultiPoint</a>
      directive.
    </div>

<h2 id="adsummary">Advanced Directive Summary</h2>
<table cellspacing=5>
<tr>
  <td><b>Input And Output</b></td>
  <td><b>Trait Model</b></td>
  <td><b>Options</b></td>
</tr>
<tr>
  <td valign='top'>
    <a href='#countfile'>CountFile</a><br>
    <a href='#forcebrfile'>ForceBRFile</a><br>
    <a href='#nidetailfile'>NIDetailFile</a><br>
    <a href='#surfacefile'>SurfaceFile</a><br>
    <a href='#surfacespath'>SurfacesPath</a><br>
  </td>
  <td valign='top'>
    <a href='#alpha'>Alpha</a><br>
    <a href='#constraint'>Constraint</a><br>
    <a href='#degreesoffreedom2'>DegreesOfFreedom</a><br>
    <a href='#diseasegenefrequency'>DiseaseGeneFrequency</a><br>
    <a href='#dprime'>DPrime</a><br>
    <a href='#femaletheta'>FemaleTheta</a><br>
    <a href='#fixedmodels'>FixedModels</a><br>
    <a href='#maletheta'>MaleTheta</a><br>
    <a href='#mean'>Mean</a><br>
    <a href='#penetrance'>Penetrance</a><br>
    <a href='#standarddev'>StandardDev</a><br>
    <a href='#theta'>Theta</a><br>
    <a href='#threshold'>Threshold</a><br>
    <a href='#truncate'>Truncate</a><br>
  </td>
  <td valign='top'>
    <a href='#maxiterations'>MaxIterations</a><br>
  </td
</tr>
</table>

<h2 id='adreference'>Advanced Directive Reference</h2>
    <div class='hang' id='alpha'>
      <b>Alpha</b><br>
      <tt>Alpha [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of alpha values that will be considered during the analysis.
      This directive requires the <a href='#fixedmodels'>FixedModels</a> directive.
      This directive is incompatible with the <a href='#markertomarker'>MarkerToMarker</a>
      directive.
    </div><p>

    <div class='hang' id='constraint'>
      <b>Constraint</b><br>
      <tt>Constraint Penetrance [ [ DD | Dd | dD | dd ] { <i>class</i> } [ == | != | > | >= ] [ DD | Dd | dD | dd ] { <i>class</i> } ] {, ... }</tt><br>
      <tt>Constraint Mean [ [ DD | Dd | dD | dd ] { <i>class</i> } [ == | != | > | >= ] [ DD | Dd | dD | dd ] { <i>class</i> } ] {, ... }</tt><br>
      <tt>Constraint StandardDev [ [ DD | Dd | dD | dd ] { <i>class</i> } [ == | != | > | >= ] [ DD | Dd | dD | dd ] { <i>class</i> } ] {, ... }</tt><br>
      <tt>Constraint DegreesOfFreedom [ [ DD | Dd | dD | dd ] { <i>class</i> } [ == | != | > | >= ] [ DD | Dd | dD | dd ] { <i>class</i> } ] {, ... }</tt><br>
      <tt>Constraint Threshold [ <i>class</i> [ == | != | > | >= ] <i>class</i> ] {, ... }</tt><br>

      This directive allows the values specified using the various
      <a href='#adsummary'>trait model directives</a> to be constrained based on trait
      genotype or liability class. For example, the trait penetrances might be constrained
      such that the penetrance given a single copy of the disease gene is always less
      than or equal to the penetrance given two copies of the disease gene. For 
      constraints based solely on trait genotypes (no liability class is specified), the
      genotypes should be different. Constraints involving a liability class require
      that the <a href='#liabilityclass'>LiabilityClass</a> directive. Constraints
      on a given dimension of the trait model (<a href='#penetrance'>Penetrance</a>,
      <a href='#Mean'>Mean</a>, <a href='#standarddev'>StandardDev</a>,
      <a href='#degreesoffreedom'>DegreesOfFreedom</a>, <a href='#threshold'>Threshold</a>)
      require that the respective directive be specified, with all the requirements
      those directives imply. Multiple constraints on the same dimension may appear,
      separated by a comma, on the same line. In this case, if either of conditions is
      true, the constraint is satisfied. Constraints that appear on separate lines
      will all be enforced. This directive requires <a href='#fixedmodels'>FixedModels</a>.
    </div><p>

    <div class='hang' id='countfile'>
      <b>CountFile</b><br>
      <tt>CountFile <i>filename</i></tt><br>
      Specifies the name of the input file from which pedigree count information will
      be read. The default is that pedigrees are not counted. Pedigree counting can 
      dramatically improve efficiency by skipping redundant calculations when multiple
      families have the exact same structure, genotypes and affection status. This makes
      it very useful for case/control, trio or affected sib-pair analyses.
    </div><p>

    <div class='hang' id='degreesoffreedom2'>
      <b>DegreesOfFreedom</b><br>
      <tt>DegreesOfFreedom [ DD | Dd | dD | dd ] [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of degrees of freedom values to be considered for each trait
      genotype (<tt>DD</tt>, <tt>Dd</tt>, <tt>dD</tt> or <tt>dd</tt>), when using
      the Chi-squared distribution for a quantitative trait analysis. Without the
      <a href='#fixedmodels'>FixedModels</a> directive, this directive follows the
      form described in the basic <a href='#degreesoffreedom'>Directive Reference</a>.
      Otherwise, this directive requires either the <a href='#qt'>QT</a> or
      <a href='#qtt'>QTT</a> directives with the <tt>ChiSq</tt> distribution. The
      <tt>dD</tt> genotype is only legal if the <a href=#'imprinting'>Imprinting</a>
      directive is also specified. This directive is incompatible with the
      <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='diseasegenefrequency'>
      <b>DiseaseGeneFrequency</b><br>
      <tt>DiseaseGeneFrequency [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of disease gene frequency values that will be considered during
      the analysis. This directive requires the <a href='#fixedmodels'>FixedModels</a>
      directive, and is incompatible with the <a href='#markertomarker'>MarkerToMarker</a>
      directive.
    </div><p>

    <div class='hang' id='dprime'>
      <b>DPrime</b><br>
      <tt>DPrime [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of D' values that will be considered during the analysis.
      This directive requires the <a href='#fixedmodels'>FixedModels</a> and
      <a href='#ld'>LD</a> directives, and is not compatible with the
      <a href='#multipoint'>MultiPoint</a> directive.
    </div><p>

    <div class='hang' id='femaletheta'>
      <b>FemaleTheta</b><br>
      <tt>FemaleTheta [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of female theta values that will be considered during
      the analysis. This directive requires the <a href='#fixedmodels'>FixedModels</a>
      and <a href='#sexspecific'>SexSpecific</a> directives.
    </div><p>

    <div class='hang' id='fixedmodels'>
      <b>FixedModels</b><br>
      <tt>FixedModels</tt><br>
      Specifies that Kelvin should not sample the trait space using numerical
      techniques, but should calculate Bayes ratios at specific points in the
      trait space as specified using <a href='#adsummary'>trait model directives.</a>
    </div><p>

    <div class='hang' id='forcebrfile'>
      <b>ForceBRFile</b><br>
      <tt>ForceBRFile</tt><br>
      Specifies that a <a href='bayesratiofile'>BayesRatioFile</a> be written
      during a <a href='#markertomarker'>MarkerToMarker</a> analysis. This option is 
      for debugging purposes.
    </div><p>

    <div class='hang' id='maletheta'>
      <b>MaleTheta</b><br>
      <tt>MaleTheta [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of male theta values that will be considered during the
      analysis. This directive requires the <a href='#fixedmodels'>FixedModels</a>
      and <a href='#sexspecific'>SexSpecific</a> directives.
    </div><p>

    <div class='hang' id='maxiterations'>
      <b>MaxIterations</b><br>
      <tt>MaxIterations <i>number</i></tt><br>
      Specifies that Kelvin's dynamic trait space sampling algorithm should go through
      at most <tt><i>number</i></tt> iterations.
    </div><p>

    <div class='hang' id='mean'>
      <b>Mean</b><br>
      <tt>Mean [ DD | Dd | dD | dd ] [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of mean values to be considered for each trait genotype
      (<tt>DD</tt>, <tt>Dd</tt>, <tt>dD</tt> or <tt>dd</tt>), when using the normal
      distribution for a quantitative trait analysis. This directive requires the
      <a href='#fixedmodels'>FixedModels</a> directive, and either the <a href='#qt'>QT</a>
      or <a href='#qtt'>QTT</a> directives with the <tt>Normal</tt> distribution. The
      <tt>dD</tt> genotype is only legal if the <a href='#imprinting'>Imprinting</a>
      directive is also specified. This directive is incompatible with the
      <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='nidetailfile'>
      <b>NIDetailFile</b><br>
      <tt>NIDetailFile <i>filename</i></tt><br>
    </div><p>

    <div class='hang' id='penetrance'>
      <b>Penetrance</b><br>
      <tt>Penetrance [ DD | Dd | dD | dd ] [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of penetrances to be considered for each trait genotype
      (<tt>DD</tt>, <tt>Dd</tt>, <tt>dD</tt> or <tt>dd</tt>), when conducting a
      dichotomous trait analysis. This directive requires the
      <a href='#fixedmodels'>FixedModels</a>, and is incompatible with the
      <a href='#qt'>QT</a>, <a href='#qtt'>QTT</a> and
      <a href='#markertomarker'>MarkerToMarker</a> directives. The <tt>dD</tt>
      genotype is only legal if the <a href='#imprinting'>Imprinting</a> directive
      is also specified.
    </div><p>

    <div class='hang' id='standarddev'>
      <b>StandardDev</b><br>
      <tt>StandardDev <i>min</i>, <i>max</i></tt><br>
      <tt>StandardDev [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of standard deviation values to be considered for each trait genotype
      (<tt>DD</tt>, <tt>Dd</tt>, <tt>dD</tt> or <tt>dd</tt>), when using the normal
      distribution for a quantitative trait analysis. This directive requires the
      <a href='#fixedmodels'>FixedModels</a> directive, and either the <a href='#qt'>QT</a>
      or <a href='#qtt'>QTT</a> directives with the <tt>Normal</tt> distribution. The
      <tt>dD</tt> genotype is only legal if the <a href='#imprinting'>Imprinting</a>
      directive is also specified. This directive is incompatible with the
      <a href='#markertomarker'>MarkerToMarker</a> directive.
    </div><p>

    <div class='hang' id='surfacefile'>
      <b>SurfaceFile</b><br>
      <tt>SurfaceFile <i>filename</i></tt><br>
    </div><p>

    <div class='hang' id='surfacespath'>
      <b>SurfacesPath</b><br>
      <tt>SurfacePath <i>dirname</i></tt><br>
    </div><p>

    <div class='hang' id='theta'>
      <b>Theta</b><br>
      <tt>Theta [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of sex-averaged thetas to be considered during the analysis.
      This directive requires <a href='#fixedmodels'>FixedModels</a>, and is
      incompatible with the <a href='#sexspecific'>SexSpecific</a> directive.
    </div><p>

    <div class='hang' id='threshold'>
      <b>Threshold</b><br>
      <tt>Threshold [ <i>value</i> | <i>range</i> ] {, ... }</tt><br>
      Specifies the set of threshold <a href='#arguments'>values</a> for use
      with the <a href='#qtt'>QTT</a> directive. This directive requires the
      <a href='#qtt'>QTT</a> and <a href='#fixedmodels'>FixedModels</a> directives.
      Note that without the <a href='#fixedmodels'>FixedModels</a> directive,
      this directive may be specified with only two values (minimum and 
      maximum).
    </div><p>

    <div class='hang' id='truncate'>
      <b>Truncate</b><br>
      <tt>Truncate [ Left <i>value</i> | Right <i>value</i> ] {, ... }</tt><br>
    </div><p>

  </body>
</html>
