<html>
<head><title>kelvin Configuration File</title></head>
<body>
<h1>kelvin Configuration File</h1>

<a href="#analysis">Types of Analysis</a><br>
<a href="#summary">Directive Summary</a><br>
<a href="#reference">Directive Reference</a><br>
<a href="#ranges">Specifying Ranges</a><br>
<hr>

<h2 id="analysis">Types of Analysis</h2>
While most combinations of analysis type directives are valid, not all
make sense, so it is helpful to have a list of valid combinations. The specific
directives that are required for each type of analysis are provided in parenthesis.
<ul>
<li>Trait-to-marker
<ul>
<li>Multipoint sex-averaged
<ul>
<li>Dichotomous trait (<a href="#SA">SA</a>, <a href="#DT">DT</a>)
<li>Quantitative trait
<ul>
<li>Chi-square distribution (<a href="#SA">SA</a>, <a href="#QT">QT chisq</a>)
<li>Truncated Student's t-distribution (<a href="#SA">SA</a>, <a href="#QT">QT T</a>)
</ul>
<li>Quantitative trait with threshold
<ul>
<li>Chi-square distribution (<a href="#SA">SA</a>, <a href="#CT">CT chisq</a>)
<li>Truncated Student's t-distribution (<a href="#SA">SA</a>, <a href="#CT">CT T</a>)
</ul>
</ul>
<li>Multipoint sex-specific
<ul>
<li>Dichotomous trait (<a href="#SS">SS</a>, <a href="#DT">DT</a>)
<li>Quantitative trait
<ul>
<li>Chi-square distribution (<a href="#SS">SS</a>, <a href="#QT">QT chisq</a>)
<li>Truncated Student's t-distribution (<a href="#SS">SS</a>, <a href="#QT">QT T</a>)
</ul>
<li>Quantitative trait with threshold
<ul>
<li>Chi-square distribution (<a href="#SA">SA</a>, <a href="#CT">CT chisq</a>)
<li>Truncated Student's t-distribution (<a href="#SA">SA</a>, <a href="#CT">CT T</a>)
</ul>
</ul>
<li>Two-point
<ul>
<li>Dichotomous trait
<ul>
<li>Linkage equilibrium (<a href="#TP">TP</a>, <a href="#DT">DT</a>)
<li>Linkage disequilibrium (<a href="#TP">TP</a>, <a href="#DT">DT</a>, <a href="#LD">LD</a>)
</ul>
<li>Quantitative trait
<ul>
<li>Chi-square distribution
<ul>
<li>Linkage equilibrium (<a href="#TP">TP</a>, <a href="#QT">QT chisq</a>)
<li>Linkage disequilibrium (<a href="#TP">TP</a>, <a href="#QT">QT chisq</a>, <a href="#LD">LD</a>)
</ul>
<li>Truncated Student's t-distribution
<ul>
<li>Linkage equilibrium (<a href="#TP">TP</a>, <a href="#QT">QT T</a>)
<li>Linkage disequilibrium (<a href="#TP">TP</a>, <a href="#QT">QT T</a>, <a href="#LD">LD</a>)
</ul>
</ul>
<li>Quantitative trait with threshold
<ul>
<li>Chi-square distribution
<ul>
<li>Linkage equilibrium (<a href="#TP">TP</a>, <a href="#CT">CT chisq</a>)
<li>Linkage disequilibrium (<a href="#TP">TP</a>, <a href="#CT">CT chisq</a>, <a href="#LD">LD</a>)
</ul>
<li>Truncated Student's t-distribution
<ul>
<li>Linkage equilibrium (<a href="#TP">TP</a>, <a href="#CT">CT T</a>)
<li>Linkage disequilibrium (<a href="#TP">TP</a>, <a href="#CT">CT T</a>, <a href="#LD">LD</a>)
</ul>
</ul>
</ul>
</ul>
<li>Marker-to-marker
<ul>
<li>Linkage equilibrium (<a href="#MM">MM</a>)
<li>Linkage disequilibrium (<a href="#MM">MM</a>, <a href="#LD">LD</a>)
</ul>
</ul>
<hr>

<h2 id="summary">Directive Summary</h2>

<h4>Analysis Type</h4>
<ul>
<li><a href="#SS">SS &lt;m&gt;</a>: sex-specific multipoint analysis with &lt;m&gt; markers.
<li><a href="#SA">SA &lt;m&gt;</a>: sex-averaged multipoint analysis with &lt;m&gt; markers.
<li><a href="#TP">TP</a>: two-point analysis.
<li><a href="#DT">DT</a>: dichotomous (binary) trait analysis.
<li><a href="#QT">QT</a>: quantitative trait analysis, specifying distribution type and values.
<li><a href="#CT">CT</a>: quantitative trait analysis with threshold (combined), specifying distribution type and values.
<li><a href="#MM">MM</a>: marker-to-marker analysis as opposed to the default trait-to-marker analysis.
<li><a href="#LD">LD</a>: linkage disequilibrium analysis as opposed to the default linkage equilibrium,
specifying multiple D' values or a range.
</ul>
<h4>File Specification</h4>
<ul>
<li><a href="#HE">HE &lt;file name&gt;</a>: bayes ratio output file, default bf.out.
<li><a href="#PF">PF &lt;file name&gt;</a>: posterior probability of linkage output file, default ppl.out.
<li><a href="#PD">PD &lt;file name&gt;</a>: pedigree input file, default pedfile.dat.
<li><a href="#DF">DF &lt;file name&gt;</a>: pedigree locus description data input file, default datafile.dat.
<li><a href="#MK">MK &lt;file name&gt;</a>: marker input file, default markers.dat.
<li><a href="#MP">MP &lt;file name&gt;</a>: map input file, default mapfile.dat.
<li><a href="#CF">CF &lt;file name&gt;</a>: case control input file.
<li><a href="#SR">SR &lt;path&gt;</a>: saved results input/output path.
</ul>
<h4>Data Values</h4>
<ul>
<li><a href="#UP">UP &lt;n&gt;</a>: unknown person ID, default "0".
<li><a href="#AS">AS &lt;n&gt; &lt;m&gt; &lt;l&gt;</a>: affection status values
</ul>
<h4>Multipoint Analysis Only</h4>
<ul>
<li><a href="#TL">TL</a>: trait locations, specifying multiple values or a range.
<li><a href="#TM">TM</a>: trait loci that map exactly to markers are included in tested trait loci.
<li><a href="#AL">AL</a>: alpha parameter values, specifying multiple values or a range.
</ul>
<h4>Two-Point Analysis Only</h4>
<ul>
<li><a href="#Th">Th</a>: theta, specifying multiple sex-averaged values or a range.
<li><a href="#Tm">Tm</a>: theta, specifying multiple sex-specific male values or a range.
<li><a href="#Tf">Tf</a>: theta, specifying multiple sex-specific female values or a range.
</ul>
<h4>Quantitative Trait with Threshold Only</h4>
<ul>
<li><a href="#TT">TT</a>: 
<li><a href="#T_MIN">T_MIN</a>: 
<li><a href="#T_MAX">T_MAX</a>: 
</ul>
<h4>Truncated Student's T-Distribution Only</h4>
<ul>
<li><a href="#MIN">MIN</a>: 
<li><a href="#MAX">MAX</a>: 
</ul>
<h4>All Types of Analyses</h4>
<ul>
<li><a href="#XC">XC</a>: analysis of X-chromosome instead of default autosome.
<li><a href="#PE">PE &lt;n&gt;</a>: polynomial evaluation with scaling &lt;m&gt; instead of iterative calculation.
<li><a href="#DD, Dd, dd">DD, Dd, dd</a>: penetrance, specifying multiple values or a range, and constraints.
<li><a href="#P1">P1</a>: used in conjunction with penetrance.
<li><a href="#GF">GF</a>: gene frequency, specifying multiple values or a range.
<li><a href="#DA">DA</a>: number of disease alleles, currently must be 2, defaults to 2.
<li><a href="#AF">AF</a>: allele frequencies for SNPs, specifying multiple values or a range.
</ul>
<h4>Other Directives</h4>
<ul>
<li><a href="#LC">LC</a>: 
</ul>
<hr>

<h2 id="reference">Directive Reference</h2>
<table border="1">
<tr>
<td width="10%">
<a href="#AF">AF</a><br>
<a href="#AL">AL</a><br>
<a href="#MM">AM/MM</a><br>
<a href="#AS">AS</a><br>
<a href="#CF">CF</a><br>
<a href="#CT">CT</a><br>
<a href="#DA">DA</a><br>
<a href="#DD, Dd, dd">DD, Dd, dd</a><br>
<a href="#DF">DF</a><br>
<td width="10%">
<a href="#DT">DT</a><br>
<a href="#GF">GF</a><br>
<a href="#HE">HE</a><br>
<a href="#LC">LC</a><br>
<a href="#LD">LD</a><br>
<a href="#MAX">MAX</a><br>
<a href="#MIN">MIN</a><br>
<a href="#MK">MK</a><br>
<a href="#MP">MP</a><br>
<td width="10%">
<a href="#P1">P1</a><br>
<a href="#PD">PD</a><br>
<a href="#PE">PE</a><br>
<a href="#PF">PF</a><br>
<a href="#QT">QT</a><br>
<a href="#SA">SA</a><br>
<a href="#SR">SR</a><br>
<a href="#SS">SS</a><br>
<a href="#TL">TL</a><br>
<td width="10%">
<a href="#TM">TM</a><br>
<a href="#TP">TP</a><br>
<a href="#TT">TT</a><br>
<a href="#T_MAX">T_MAX</a><br>
<a href="#T_MIN">T_MIN</a><br>
<a href="#Tf/Tm">Tf/Tm</a><br>
<a href="#Th">Th</a><br>
<a href="#UP">UP</a><br>
<a href="#XC">XC</a><br>
</tr>
</table>

<h3 id="AF">AF</h3>

<h3 id="AL">AL</h3>

<h3 id="AS">AS</h3>
Specifies the values used to represent affection status in the pedigree data file.
<pre>
   AS 0 1 2
</pre>
where the first value is the "unknown" value, the second value is the
"unaffected" value, and the third value is the "affected" value (the
numbers shown reflect the defaults for DT; for multipoint, the
defaults are NAN, -9999.99, and 9999.99, respectively).

<h3 id="CF">CF</h3>

<h3 id="CT">CT</h3>

<h3 id="DA">DA</h3>

<h3 id="DD, Dd, dd">DD, Dd, dd</h3>

<h3 id="DF">DF</h3>

<h3 id="DT">DT</h3>

<h3 id="GF">GF</h3>

<h3 id="HE">HE</h3>

<h3 id="LC">LC</h3>
We need to accomodate liability classes for all three models, and also
support constraints between liability classes. Thus liability classes
allow you to apply different penetrance models to individuals,
supporting things like, e.g., age of onset, or imprinting if you use
liability classes to code for M/F parents.
<p>
In practice, what this means is that when you have multiple liability
classes, you get "extra" dimensions for the LOD scores. Moreover, you
need to support cross-class constraints, as well as differing
within-class constraints. What you don't need to do is to support,
e.g., dichotomouse traits for one class and quantitative traits for
the other class, or different grid specifications for the different
classes.
<p>
For example, if you have two liability classes LC 1 and LC 2 in a DT
model:
<pre>
    LC 2
    DD 0.0 0.8 0.1; 0.9 0.999 0.099
    Dd 0.0 0.8 0.1; 0.9 0.999 0.099
    dd 0.0 0.8 0.1; 0.9 0.999 0.099
    Dd > dd
    DD 1 > Dd 1
    DD 1 > DD 2
</pre>
Note the extension to the constraint language; keywords without an
integer class number apply uniformly to all classes, while those with
integer class numbers provide inter-class constraints or within-class
constraints applied to only select classes.
<p>
This specification says that Dd>dd for both LC 1 and LC 2, but DD>Dd
only for LC 1 and not LC 2, and LC 1's DD always exceeds LC 2's DD,
excluding grid points where this is not the case.
<p>
As before, for multiallelic diseases, we would lose the Dd
specifications and instead use single-digit integers.

<h3 id="LD">LD</h3>

<h3 id="MAX">MAX</h3>

<h3 id="MIN">MIN</h3>

<h3 id="MK">MK</h3>

<h3 id="MM">MM</h3>
The MM directive indicates marker to marker analysis. By default, it
will do all possible pairwise analyses between any two markers. In
contrast, the AM directive specifies that only adjacent markers will
be analyzed.

<h3 id="MP">MP</h3>

<h3 id="P1">P1</h3>

<h3 id="PD">PD</h3>

<h3 id="PE">PE</h3>
Applicable to all types of analyses.
<p>
Indicates that polynomial evaluation should be performed for likelihood
calculations; overrides the default explicit
evaluation condition.
<pre>
   PE n
</pre>
Where n is an optional integer parameter which specifies the polynomial
scaling factor in the range of 1-10. Default is 1, and best for
environments with limited memory resources. Higher values will generally
trade memory for execution time.

<h3 id="PF">PF</h3>
Applicable to all types of analyses.
<p>
Specifies the name of the post-MAKEPED format pedigree file to be used
in the analysis.
<pre>
   PF pedfile.dat
</pre>
Shown with default file name. The parameter can include a path which can be absolute or
relative to the current default directory. A common use of this is
specifying a single pedigree file to be shared by multiple analyses in separate
subdirectories as:
<pre>
   PF ../pedfile.dat
</pre>
For information on the format of the pedigree file itself, see <emp>TBS</emp>.

<h3 id="QT">QT</h3>
Quantitative trait analysis requires specification of a trait distribution. Currently
normal, Student's T, and Chi-squared distributions are supported.
<p>
Normal is not
recommended except with test data.
<p>
Student's T-distribution is specified as:
<tt>
QT T df mean sd
</tt>
where <tt>df</tt> is the number of degrees of freedom, e.g. 30, <tt>mean</tt> is the mean of the
trait values, and <tt>sd</tt> is their standard deviation.
<p>
Chi-squared distribution is specified as:

<h3 id="SA">SA</h3>

<h3 id="SR">SR</h3>
SR - takes an optional parameter that specifies the path in
which to save/restore intermediate results of a fixed-grid analysis.
The fixed grid must consist of 6 disease gene frequencies and 275
penetrance values, e.g.:
<pre>
GF 0.001;0.01;0.1;0.3;0.5;0.8

DD 0.0 0.9 0.1
Dd 0.0 0.9 0.1
dd 0.0 0.9 0.1
DD 0.999
Dd 0.999
dd 0.999
DD >= Dd
Dd >= dd
DD != Dd; Dd != dd
</pre>
Trait, marker, and alternative results are stored in files named:
<pre>
<optional path/>trait/ped-<pn>/ped-<pn>_trait.tpl - for 22 allosomes
<optional path/>trait-23/ped-<pn>/ped-<pn>_trait.tpl - for autosome
<optional path/>chr-<cn>_ped-<pn>/chr-<cn>_ped-<pn>_<m1>_<m2>_marker.tpl
<optional path/>chr-<cn>_ped-<pn>/chr-<cn>_ped-<pn>_pos-<po>_alternative.tpl
</pre>
where <optional path> is the parameter to the SR directive, if specified,
<i><pn></i> is pedigree number, <i><cn></i> is chromosome number, <i><m1></i> and <i><m2></i>
are marker names, and <i><po></i> is position.
<p>
Paths are created as needed. If trait, marker or alternative result files
are found, they are loaded and the corresponding calculations skipped. When
calculations are performed, trait, marker and alternative results are written.
To force recalculation, delete the saved results.

<h3 id="SS">SS</h3>

<h3 id="TL">TL</h3>

<h3 id="TP">TP</h3>

<h3 id="TT">TT</h3>

<h3 id="T_MAX">T_MAX</h3>

<h3 id="T_MIN">T_MIN</h3>

<h3 id="Tf">Tf</h3>

<h3 id="Th">Th</h3>

<h3 id="Tm">Tm</h3>

<h3 id="UP">UP</h3>

<h3 id="XC">XC</h3>
<hr>

<h2 id="ranges">Specifying Ranges</h2>
</html>
<hr>
$Date$
</body>

</html>
