<html>
  <?-- This file is 100 characters wide -->
  <head>
    <title>KELVIN V2.1.0 ReadMe</title>
    <style type='text/css'>
      span.sc { font-variant: small-caps; }
      div.indent { padding-left: 2em; }
      div.hang {padding-left: 2em ; text-indent: -2em; }
      div.footnote { font-size: smaller; }
      td.lineno { text-align: right; background-color: #888888; }
      td.colno { text-align: center; background-color: #888888; width: 2em }
    </style>
  </head>
  <body>
    <center>
    <h1><span class='sc'>Kelvin</span> V2.1.0</h1>
    </center>

    <h1>Overview</h1>

    <span class='sc'>Kelvin</span> is a flexible program for analysis of genetic data.  It is
    based on the PPL framework<sup><a href='#foot1'>1</a>, <a href='#foot2'>2</a>,
    <a href='#foot3'>3</a></sup>, and produces output on the posterior probability (0,..,1)
    scale. Primary outcome statistics include the PPL (posterior probability of linkage (L)),
    as well as various forms of the PPLD (posterior probability of linkage disequilibrium (LD)).
    In the context of analysis between a trait locus and a marker, the PPLD models allelic
    association due to very close genomic proximity. Versions of the PPLD include the PPLD(L)
    (posterior probability of L and LD); the PPL(LD) (posterior probability of L allowing for LD);
    and the PPLD|L (posterior probability of LD given L).
    <p>
    <span class='sc'>Kelvin</span> handles the trait model by formulating an underlying
    likelihood in terms of a vector of trait parameters, and integrating over all constituent
    trait parameters. The standard dichotomous trait (DT) likelihood is written in terms of a
    single disease allele frequency (for a two-allele disease locus), a vector of three
    penetrances (one for each disease genotype), and the admixture parameter of
    Smith<sup><a href='#foot4'>4</a></sup> (to allow for heterogeneity within data sets).
    Quantitative trait (QT) models are parameterized as mixtures of continuous distributions,
    one for each underlying disease genotype.  At present QT models can be configured as mixtures
    of normal distributions (three means and three variances) or as mixtures of chi-square
    distributions (three d.f. parameters). Also available is a QT threshold (QTT) model, which
    adds an additional threshold parameter to the likelihood, allowing a mix of binary
    (affected/unaffected) and QT phenotypes within the same pedigrees.
    <p>
    <span class='sc'>Kelvin</span> can accept as input multigenerational pedigree data, nuclear
    families, "trios" (one affected child with two parents), and case-control data. There is no
    need for the user to specify the data structures when running the program, and all models
    can be applied to any structure. Limitations include application to very large pedigrees
    for which the exact likelihood calculation may be intractable; and at present, multipoint
    analyses involving large numbers of markers. (At present <span class='sc'>Kelvin</span> is
    based on the Elston-Stewart algorithm<sup><a href='#foot5'>5</a></sup>; implementation of the
    Lander-Green algorithm<sup><a href='#foot6'>6</a></sup> is underway.) Of course running the
    PPL itself on case-control or trio data alone will not yield any information, because these
    structures do not contain information on recombination in the absence of LD.
    <p>
    <span class='sc'>Kelvin</span> handles potential inter-sample heterogeneity via Bayesian
    sequential updating. For this purpose, the user submits data in separate subsets. These can
    be formed based on any clinical and/or demographic information which seems likely to demarcate
    more homogeneous subsets of data. Sequential updating can and should also be used when
    considering new data, "updating" previously calculated <span class='sc'>Kelvin</span> results
    with the newly obtained data. That is, rather than adding the new data to the old data and
    recomputing <span class='sc'>Kelvin</span> results based on the 'pooled' data set,
    <span class='sc'>Kelvin</span> is designed to sequentially update previously obtained results
    with new data as these become available. <span class='sc'>Kelvin</span> then integrates over
    the trait parameters separately in each data subset, allowing the possibility that underlying
    trait parameters do not have the same values across subsets. (It is advisable to introduce
    new data in "chunks," that is, not one family at a time, but in batches
    <sup><a href='#foot7'>7</a></sup>).
    <p>
    <span class='sc'>Kelvin</span> also provides various options for allowing the pentrances
    (or in the case of QTs, means and variances or d.f.) to depend upon covariates. Any
    classification can be used for this purpose, such as age classes, sex, or clinical features
    (e.g., "strict" versus "broad" diagnoses).  To implement such an analysis the user assigns
    individuals to liability classes based on the covariate, and <span class='sc'>Kelvin</span>
    internally integrates over the penetrances separately across the classes. This feature can
    also be used to implement two-locus (2L) epistasis analyses based on measured genotypes at,
    e.g., an associated SNP, by allowing the penetrances to depend on SNP genotype.
    <span class='sc'>Kelvin</span> can also perform analyses allowing for imprinting, or
    penetrances dependent on parent of origin.
    <p>
    The <a href='#releasenotes'>Release Notes</a> describe the major changes in this version
    The <a href='#quickstart'>Quick Start</a> gives a brief introduction to preparing the input
    files, running <span class='sc'>Kelvin</span>, and graphing and interpreting the output.
    Additional technical details on the underlying statistical methods can be found in the
    following references:

    <div class='indent'>
      Vieland VJ. Bayesian linkage analysis, or: How I learned to stop worrying and love the
      posterior probability of linkage. Am J Hum Genet 63(4):947-954, 1998. 
      <p> 
      Wang K, Huang J, Vieland VJ. The consistency of the posterior probability of linkage.
      Annals Hum Genet 64(Pt 6):533-553, 2000.  
      <p>
      Vieland VJ, Wang K, Huang J. Power to detect linkage based on multiple sets of data in
      the presence of locus heterogeneity: Comparative evaluation of model-based linkage methods
      for affected sib pair data. Hum Hered 51(4):199-208, 2001.
      <p>
      Logue MW, Vieland VJ. A new method for computing the multipoint posterior probability of
      linkage. Hum Hered 57:90-99, 2004.
      <p>
      Yang X, Huang J, Logue MW, Vieland VJ. The posterior probability of linkage allowing for
      linkage disequilibrium and a new estimate of disequilibrium between a trait and a marker.
      Human Hered 59:210-219, 2005.
      <p>
      Vieland VJ. Thermometers: Something for statistical geneticists to think about. Human
      Hered, 61:144-156, 2006. 
      <p>
      Bartlett CW, Vieland VJ. Accumulating quantitative trait linkage evidence across multiple
      datasets using the posterior probability of linkage.  Genet Epidem, 31(2):91-102, 2007.
      <p>
      Wang H, Segre AM, Huang Y, O'Connell J, Vieland VJ. Fast computation of human genetic
      linkage.  Proceedings of IEEE 7th Symposium on Bioinformatics and Bioengineering (BIBE
      2007), pages 857-863, 2007.
      <p>
      Govil M, Vieland VJ. Practical considerations for dividing data into subsets prior to PPL
      analysis.  Hum Hered 66:223-237, 2008.
      <p>
      Seok S, Evans M, Vieland VJ. Fast and accurate calculation of a computationally intensive
      statistic for mapping disease genes. J Comput Biol 16(5):659-676, 2009.
    </div>

    <div class='footnote'>
      References: <sup>1</sup>Smith, C.A.B. Some comments on the statistical methods used in
      linkage investigations. Am J Hum Genet 11, 289-304 (1959). <sup>2</sup>Vieland, V.J.
      Bayesian linkage analysis, or: how I learned to stop worrying and love the posterior
      probability of linkage. Am J Hum Genet 63, 947-54 PMID: 9758634 (1998). <sup>3</sup>Vieland,
      V.J. Thermometers: something for statistical geneticists to think about. Hum Hered 61,
      144-56 PMID: 16770079 (2006). <sup>4</sup>Smith, C.A.B. Testing for heterogeneity of
      recombination fraction values in human genetics. Ann Hum Genet 27, 175-182 (1963).
      <sup>5</sup>Elston, R.C. & Stewart, J. A general model for the genetic analysis of pedigree
      data. Hum Hered 21, 523-42 (1971). <sup>6</sup>Lander, E.S. & Green, P. Construction of
      multilocus genetic linkage maps in humans. Proc Natl Acad Sci U S A 84, 2363-7 (1987).
      <sup>7</sup>Govil, M. & Vieland, V.J. Practical considerations for dividing data into
      subsets prior to PPL analysis. Hum Hered 66, 223-237 PMID: 18612207 (2008).
    </div>

    <hr>

    <h1 id='releasenotes'>Release Notes</h1>

    V2.1 is a major release of <span class='sc'>Kelvin</span> that provides significant
    improvements in capabilities and performance. Users of previous versions will want to
    review the changes in this section before installing and using the new version.
    
    <ol>
      <li>The names and formats of directives in the configuration file have changed significantly. See the
	<a href='Directives.html'>directives page</a> for a complete reference.
	The Perl script <tt>convertconfig.pl</tt> is provided to help convert existing configuration files
	to the new format. It can be invoked by typing:
	<pre>
	  _$ convertconfig.pl &lt;old.conf &gt;new.conf
	</pre>
	Be sure to review and modify the new configuration file as required, since not all aspects
	of old configurations are supported.
      </li>
      <li>Output file format has changed significantly, so if you have written
	scripts or programs to process that output, you will want to review the new formats
	in the <a href='#outputfiles'>Contents Of Output Files</a> section.</li>
      <li>The default mechanism of integration over constituent trait parameters is now adaptive 
	leading to significant improvements in performance and accuracy (see Seok, Evans
	and Vieland).</li>
      <li>Pedigree files may generally be in pre-MAKEPED format, except in limited cases,
        as detailed under <a href='#pedfileformat'>data file formats</a>.
      <li><a href='#pedcounting'>Pedigree counting and binning</a> is now performed by
        default when possible. Redundancies in the pedigree file are consolidated to reduce
        calculations. This behavior can be explicitly disabled through a configuration
        directive.</li>
      <li>Marker allele frequencies can be <a href='#frequencyestimation'>automatically
        estimated</a> for some linkage disequilibrium analyses. These frequencies
	will be used in preference to user-supplied allele frequencies. This behavior can be
        explicitly disabled through a configuration directive.</li>
      <li>Individuals can now be divided into groups with separate sets of penetrance values. 
	See the section on <a href='#lc_vs_nolc'>Covariate-Dependence</a> for details.</li>
      <li>Epistasis analysis is now supported, see the section on 
	<a href='#simple_vs_epistasis'>Single-locus vs 2L Epistasis</a> for details.</li>
      <li>Linkage results can be combined with LD results in post processing using the
        <a href='#sequpdate'>sequential updating</a> tools.
      <li>Utilities for preparing the configuration file and graphing the results of analyses
	are now provided. Separate documentation is provided for both
	<tt><font color='red'>gKelvin</font></tt> and <tt><font color='red'>graphApp</font></tt>
        as <tt>Readme.txt</tt> files in their distribution directories. These utilities
	have separate installation requirements and procedures as well.</li>
      <li>Progress reporting now provides configurable fixed-frequency updates so as to
	eliminate the output "feast or famine" effect across differing analyses. See the
	<a href='#running'>Running The Programs</a> section for details.</li>
    </ol>

    <hr>

    <h1 id='quickstart'>Quick Start</h1>

    This section provides an overview of getting, building and running
    <span class='sc'>Kelvin</span>. Detailed instructions on each point can be found
    <a href='#toc'>below</a>.

    <h3>Installing <span class='sc'>Kelvin</span></h3>

    These instructions assume that <span class='sc'>Kelvin</span> is being installed on a Linux or
    Unix system of recent vintage, with a GCC compiler and that the user is familiar with the
    command line environment on that system. First, unpack the distribution archive with 
    <tt>gunzip</tt> and <tt>tar</tt>, and move to the newly-created <tt>kelvin-&lt;version&gt;</tt> directory.
    If you'd like to try to use one of the prebuilt  executables, simply type <tt>make install-prebuilt</tt>, 
    and if you don't receive any errors, you're done. If no prebuilt executable was available, or you want to 
    build your own version anyhow, follow these steps:
    <ul>
      <li>Edit the <tt>Makefile</tt> if desired, and then build <span class='sc'>Kelvin</span> by running <tt>make</tt>.</li>
      <li>Validate the executable by running <tt>make check</tt>. These tests should take less than a minute to perform.</li>
      <li>You can perform a much more extensive set of tests if you wish by running <tt>make test</tt>. These tests 
	will take anywhere from 30 minutes to several hours. You can interrupt them at any time by
	typing <tt>CTRL/C</tt>.</li>
      <li>Install the executable and supporting scripts by running <tt>make install</tt>.</li>
    </ul>
    Note that the default installation directory is <tt>/usr/local/bin</tt>, which will require the installer to
    have administrative privileges. To install in an alternative directory, e.g. <tt>~/kelvin</tt> for 
    individual use, edit <tt>Makefile</tt> and replace the line <tt>BINDIR=/usr/local/bin</tt> with 
    <tt>BINDIR=~/kelvin</tt> or some other path, and then replace the lines <tt>GROUP=root</tt> and <tt>USER=root</tt>
    with your group and user name before issuing the <tt>make install</tt> command.

    <h3>Preparing Input Files</h3>

    Kelvin requires four input data files. In these examples, we show an affected sib-pair
    family and three markers:

    <ol>
      <li>Pedigree File - This is a pedigree file containing phenotypic and genotypic
        information. It may generally be in either pre- or
        post-<a href='http://linkage.rockefeller.edu/soft/linkage/sec2.7.html'>MAKEPED</a>
	format, although in <a href='#pedfileformat'>specific cases</a>, post-MAKEPED format
        is required.
	<pre>
        1 1 0 0 3 0 0 1 1 1 2 2 1 2 1 1  Ped: 1  Per: 1
        1 2 0 0 3 0 0 2 0 1 1 1 1 2 1 1  Ped: 1  Per: 2
        1 3 1 2 0 4 4 2 0 2 1 2 2 2 1 1  Ped: 1  Per: 3
        1 4 1 2 0 0 0 1 0 2 2 1 1 1 1 1  Ped: 1  Per: 4
        </pre>
      </li>
      <li>Locus (Data) File - Describes column order in the pedigree file starting with 
        the position of the trait locus.
	<pre>
        T Trait
        M MRK_1
        M MRK_2
        M MRK_3
	</pre>
      </li>
      <li>Frequency File - Names the markers listed in datafile.dat and enumerates the allele
        frequencies.
        <pre>
        M MRK_1
        F 0.3 0.7
        M MRK_2
        F 0.35 0.65
        M MRK_3
        F 0.7 0.3
        </pre>
      </li>
      <li>Map File - Gives the chromosomal position of the markers.
        <pre>
        CHROMOSOME	MARKER		POSITION
        3		MRK_1		0.33
        3		MRK_2		0.66
        3		MRK_2		0.99
        </pre>
      </li>
    </ol>

    <h3>Creating A Configuration File</h3>

    There are two ways of creating a configuration file. If you are using a xterm or other X-windows-compatible
    mechanism to access your system, you can use gKelvin, which is a simple graphic interface that will lead
    you thru the necessary steps and ensure that your configuration is syntactically correct and internally
    consistent. If you cannot or do not wish to use gKelvin, you can prepare your configuration file by hand
    with any text editor.
<p>
    The configuration file is generated from <tt><font color='red'>gKelvin</font></tt>. The
    GUI provides options for different types of analyses and will write out <tt>kelvin.conf</tt>.
    After starting <tt>gKelvin</tt>, open a new file. You may then <a href='gKelvin-1.png'>select
    "Two point" or "Multi-point"</a>. Based on that choice, you will be presented with forms
    for setting <a href='gKelvin-2.png'>two-point options</a>, or
    <a href='gKelvin-3.png'>multipoint options</a>. When you've finished,
    <a href='gKelvin-4.png'>save the configuation file.</a>

    <h3>Creating a Configuration File</h3>

    Kelvin takes all of its configuration information from a single file specified on the
    command line. This file is composed of directives that describe the analysis to be performed,
    and the locations of supporting data files. The graphical configuration tool
    <font color='red'>gKelvin</font> will create the configuration for you. 
    <p>
    Kelvin supports a wide variety of analyses. The principle options which can be selected in
    gKelvin are listed below. With certain exceptions, the user is free to select any combination
    of options. gKelvin will also indicate when options are not available in combination with one
    another.
    <div class='indent'>
      <div class='hang'>
        <b>Two-point vs. multipoint analysis</b><br>
        For linkage analyses, the user can select either two-point or multipoint; for LD analyses,
        only the two-point option is available. In any case, the user can choose to use either
        sex-averaged or sex-specific marker maps.
        <p>
        <b>Marker-marker analysis</b><br>
        This option is available only for two-point analyses, but can be used to estimate either
        linkage or LD distances between pairs of markers. The user can opt to consider all
	possible pairs of markers or only pairs which are adjacent to one another in the map file.
        <p>
        <b>Linkage Disequilibrium (LD, or association) analysis</b><br>
        Unless specifically instructed to allow for LD, Kelvin assumes linkage equilibrium between
        the trait and the marker or genomic location. For two-point analyses, it is possible to
        allow for trait-marker LD.
        <p>
        <b>Trait models</b><br>
        Kelvin handles dichotomous traits (DT) or quantitative traits (QT). For QT the user can
        select the functional form of the genotypic trait distribution as either normal or
        chi-square. In addition, a QT threshold model (QTT) can be specified. In all cases Kelvin
        internally integrates over the parameters of the trait model, so there is no need to
        specify trait values or a mode of inheritance.
        <p>
        <b>Imprinting</b><br>
        The imprinting option allows penetrances (or the corresponding QT parameters) to depend on
        parent of origin.<p>
        <b>Covariate-Dependent Penetrances (or corresponding QT parameters)</b><br>
        When the user assigns individuals in a pedigree to "liability classes," Kelvin allows the
        penetrances (or QT parameters) to depend upon class assignment.
        <p>
        <b>Two-Locus (2L) Epistasis</b><br>
        Kelvin has special routines for handling two-locus analysis when supplied with genotypes at
        one of the loci (Locus 1), by allowing the penetrances (or corresponding QT parameters) to
	depend upon genotypes at Locus 1 when considering linkage and/or LD to other loci.
      </div>
    </div>

    <h3>Running <span class='sc'>Kelvin</span></h3>

    Run <tt>kelvin <i>configfile</i></tt>. The text that <span class='sc'>Kelvin</span> emits
    as it runs contains important information, and is necessary to identify the cause if the
    run fails.

    <h3>Sequential Updating</h3>

    One of <span class='sc'>Kelvin</span>'s distinctive features is the way in which it is
    designed to handle the combined analysis of multiple data sets or data subsets, using
    sequential updating. This is done by calling
    <tt><a href='#calcupdatedppl'>calc_updated_ppl</a></tt>. This program takes as input the
    output from <span class='sc'>Kelvin</span> analyes (in particular, it uses the <tt>br.out</tt>
    files). Sequential updating works with any <span class='sc'>Kelvin</span> options, however,
    it is important that all files being used as input to <tt>calc_updated_ppl</tt> should
    reflect the same sets of options. Users interested in running calc_updated_ppl on output
    generated by older versions of Kelvin must use the <tt>br.out</tt> conversion script
    <a href='#convertbr'>convert_br.pl</a>.

    <h3>Graphing The Results</h3>

    To run the graphing application, at a command prompt, goto the
    directory where the source files are
    located, and type <br>
    <br>
    <span style="font-family: monospace;">python graphApp.py</span><br>
    <br>
    <br>To import data from a <span class='sc'>Kelvin</span> output file:<br>
    1. Click on File-&gt;Import Data on the menu bar. 
    Alternatively you can click on the Import Text icon on the toolbar&nbsp;
    <img style="width: 47px; height: 42px;" alt="" src="graphApp_import_text.png"> .<br>
    <br>
    2. A dialog box will appear asking for the name of the file to be opened.
    Choose the file containing the data, and click on the Open button.<br>
    <br>
    3. Another dialog box will appear, that asks about the specific format
    of the file. A screenshot of this dialog box is shown below. At
    the very top a preview of the file showing the first five lines of the
    file is shown.<br>
    <br>
    <br>
    <div style="text-align: center;">
        <img style="width: 712px; height: 559px;" alt="" src="graphApp_fileformat_dialog.png"><br>
    </div>
    <br>
    <br>
    4. Specify the Position column in the first drop down menu box.&nbsp;
    The values of this column will be used as the x-value of points.<br>
    <br>
    5. Specify the PPL column in the second drop down menu box. The values
    of this column will be used as the y-value of points.<br>
    <br>
    6. Specify the number of header lines at the top of the file.&nbsp;
    Those lines will be skipped when processing the file.<br>
    <br>
    7. Select whether the file contains data for just one chromosome, or
    multiple chromosomes. If Single Chromosome is chosen, then the number
    of that particular chromosome needs to be specified. If Multiple
    Chromosomes are chosen, and if there is a column in the file that
    indicates chromosome number, then that needs to be specified. If no
    column indicates chromosome number, then be sure to select 'None'.<br>
    <br>
    8. Once all the required information about the file is entered, press
    the 'OK' button. The data in the file will be graphed.<br>
    <br>
    For more information about the features of the graphing application,
    please consult the manual.<br>
    <br>
    <h3>Examining The Results In Detail</h3>

    <span class='sc'>Kelvin</span> routinely creates multiple output files. For multipoint
    analyses, there is a single primary output file, the <a href='#br-multipoint'>Bayes Ratio</a>
    (BR) file, which contains the BRs for all positions at which calculations were done
    (and/or at all markers), and the PPLs themselves.  The BRs are used for sequential updating
    across data sets or subsets.  For two-point analyses, two separate files are created: the
    <a href='#br-twopoint'>BR</a> file, with the BRs; and a separate
    <a href='#ppl-twopoint'>PPL</a> file, with the PPLs.  The user can also request the MOD
    score and the maximizing values of all parameters, which are contained in the
    <a href='#modfile'>maximizing model</a> (MOD) file.

    <hr>

    <h1 id='toc'>Detailed Instructions</h1>

    <ol>
      <li><a href='#installing'>Installing The Programs</a></li>
      <li><a href='#prepconfig'>Preparing The Configuration File</a></li>
      <li><a href='#fileformats'>Input Data File Formats</a></li>
      <li><a href='#running'>Running The Programs</a></li>
      <li><a href='#outputfiles'>Contents Of Output Files</a></li>
      <li><a href='#sequpdate'>Sequential Updating</a></li>
    </ol>

    <h2 id='installing'>Installing The Programs</h2>

    <h3>Requirements</h3>

    <h4>Platforms</h4>

    <span class='sc'>Kelvin</span> is written in ANSI C, with some supporting scripts in Perl
    and Python. <span class='sc'>Kelvin</span> can generally be expected to run on any relatively
    recent Linux or Unix distribution. It has been successfully built with various versions of
    the <a href='http://gcc.gnu.org'>Gnu C Compiler (GCC)</a> and the Intel C Compiler (ICC).
    <span class='sc'>Kelvin</span> releases are currently tested on the following platforms:
    <ul>
      <li>Redhat Enterprise Linux x86_64 with GCC 4.1.1</li>
      <li>Debian Linux i386 with GCC 4.1.1</li>
      <li>Macintosh OSX 10.6 i686 with GCC 4.2.1 and GCC 4.3.0</li>
      <li>Macintosh OSX 10.4 PowerPC with GCC 4.0.1 (no OpenMP)</li>
      <li>Cygwin under Windows XP i386 with GCC 4.3.4</li>
      <li>SunOS 5.8 on an Ultra-1 sparc 64-bit with GCC 3.4.6 (no OpenMP)</li>
    </ul>
    Aside from the compiled execuatble, running <span class='sc'>Kelvin</span> requires
    <a href='http://www.perl.org'>Perl 5.x</a>. <span class='sc'>Kelvin</span>'s graphical
    configuration file generator, <font color='red'>gKelvin</font>, requires
    <a href='http://www.python.org'>Python 2.5</a> and <a href='http://www.wxpython.org'>wxPython
    2.8.3.0</a>. <span class='sc'>Kelvin</span>'s results graphing tool,
    <font color='red'>graphApp</font>, requires Python 2.5, wxPython 2.8.3.0,
    <a href='http://matplotlib.sourceforge.net'>Matplotlib</a>,
    <a href='http://numpy.scipy.org'>NumPy</a>,
    <a href='http://www.pythonware.com/products/pil'>Python Imaging Library (PIL)</a> and PyXML.

    <h4>Libraries</h4>

    Depending on the nature of the analysis, <span class='sc'>Kelvin</span> may make very
    extensive use of memory management, and can, under most circumstances, definitely benefit
    from a drop-in allocator such as <a href="http://www.hoard.org/">Hoard</a> or
    <a href="http://www.malloc.de/en/">ptmalloc3</a>. Either of these can easily halve execution
    time, and will keep memory fragmentation down when running in multi-threaded mode, but they
    are not required. For more details on using drop-in allocators with
    <span class='sc'>Kelvin</span>, see <a href="Allocators.html">the allocators section</a>.
    <p>
    When analysing quantitative traits, <span class='sc'>Kelvin</span> can also benefit from
    the statistical distribution routines in the <a href='http://www.gnu.org/software/gsl'>GNU 
    Scientific Library</a> (GSL). GSL's routines are a little faster than
    <span class='sc'>Kelvin</span>'s internal routines, and are also thread-safe.

    <h4>Resources</h4>

    <span class='sc'>Kelvin</span>'s resource requirements vary widely depending upon the
    complexity of the analysis being performed. By default, <span class='sc'>Kelvin</span>
    uses polynomials to evaluate the likelihood calculation. This generally much faster than
    not using polynomials, sometimes orders of magnitude faster, but complex pedigrees, untyped
    founders, etc., can easily push the required memory to 16GB or more. Polynomial evaluation
    can be disabled, but due to increased run times, this should only be done if limited memory
    becomes and issue. <span class='sc'>Kelvin</span> monitors CPU and memory utilization and will
    display warnings if thrashing memory.
    <p>
    If you have access to a high-performance solid-state drive (SSD), you may be able to use
    <a href="SSDSupport.html">SSD support</a> to build polynomials that are much larger than
    physical memory.
    <p>
    If you have a homogenous multi-machine Linux/GCC environment such as a Beowulf cluster and
    are performing long-running analyses, you may be able to benefit from using
    <a href="compiledPolys.html">compiled polynomials</a>, which can provide an order of magnitude
    improvement in evaluation performance, and a degree of reusability.
    <p>
    If you are running on a multi-core platform and your compiler supports
    <a href='http://www.openmp.org'>OpenMP</a> 2.0, you can build <span class='sc'>Kelvin</span> to use
    multiple threads and reduce your single-analysis runtime significantly at a marginal increase
    in memory, less if you use a drop-in allocator as described above. Note that because
    <span class='sc'>Kelvin</span>'s internal statistcal distribution routines are not
    thread-safe, building <span class='sc'>Kelvin</span> with OpenMP requires GSL.

    <h3>Distribution</h3>

    <span class='sc'>Kelvin</span> is developed in unix environments, so standard unix tools are
    used for build management and distribution.
    <p>
    <span class='sc'>Kelvin</span> is currently distributed as a compressed tarball of the
    directory hierarchy of our latest release, e.g. <tt>kelvin-2.1.0.tar.gz</tt>.

    <h4>Directory Structure and Components</h4>

    <ul>
      <li><tt>&lt;RELEASE&gt;/</tt> - <span class='sc'>Kelvin</span> source components.
        Includes normal and debug versions of the master Makefile. <span class='sc'>Kelvin</span>
	is the primary analysis program and the focus of this documentation.</li>
      <li><tt>&lt;RELEASE&gt;/bin/</tt> - simple prebuilt <span class='sc'>Kelvin</span> binaries
        for different platforms. One of these may work for you.</li>
      <li><tt>&lt;RELEASE&gt;/doc/</tt> - (very) limited documentation, including these HTML
        files.</li>
      <li><tt>&lt;RELEASE&gt;include/</tt> - a work directory for builds.</li>
      <li><tt>&lt;RELEASE&gt;/lib/</tt> - a work directory for builds.</li>
      <li><tt>&lt;RELEASE&gt;pedlib/</tt> - pedigree, likelihood, and polynomial components.
        Includes a dependent Makefile - don't try to use it by itself.</li>
      <li><tt>&lt;RELEASE&gt;/seq_update/</tt> - routines to perform a sequential updates of the
        results of previous <span class='sc'>Kelvin</span> analyses.</li>
      <li><tt>&lt;RELEASE&gt;/test-suite/</tt> - analyses that can be used for build validation
        as well as configuration examples.</li>
      <li><tt>&lt;RELEASE&gt;/utils/</tt> - general-purpose utility components. Includes a
        dependent Makefile.</li>
    </ul>

    <h4 id="building">Building and Testing</h4>

    If you cannot or do not wish to build a version of <span class='sc'>Kelvin</span> specific
    to your computing platform, you might be able to run one of the binary versions provided in
    the distribution. Binary versions of <span class='sc'>Kelvin</span> are statically-linked
    executables built without support for multi-threading. Version are provided for each of the
    environments for which we can build and test, and are in the <tt>bin</tt> subdirectory of
    the distribution. Simply copy the version you will be using to a directory on your PATH, and
    either name it <span class='sc'>Kelvin</span> or create a link to it named
    <span class='sc'>Kelvin</span>.
    <p>
    There are a number of compilation conditionals that can be set in the Makefile to affect the
    behavior of <span class='sc'>Kelvin</span>. These are disabled when commented-out with a
    pound-sign (#) at the start of the line, and enabled otherwise. Some of the less useful
    diagnostic conditionals will not be covered here.

    <ul>
      <li><tt>USE_OPENMP</tt> Enables OpenMP multi-threading. This must be commented-out if your
        compiler does notsupport OpenMP. This is enabled in the Makefile by default, but disabled
        in the prebuilt binaries. The number of threads used is determined from the environment
        variable <tt>OMP_NUM_THREADS</tt>. If this variable is not set or is set to 0 (zero) the
        behavior depends upon the operating system, but typically the number of threads used will
        correspond to the number of cores available. If <tt>OMP_NUM_THREADS</tt> is set to 1
        (one), then only a single thread is used and it is as if OpenMP multi-threading is
        disabled. OpenMP multi-threading can greatly improve polynomial evaluation, but at a
        moderate memory cost. The thread count should be less than the number of pedigrees in
        your analysis, up to the maximum of the number of cores on your platform. Monitoring
        user and system mode CPU utilization with <tt>ps</tt> during polynomial evaluation will
        be your best indicator of the efficiency of threading in your situation, and should
        guide you in selecting the right number of threads. Obviously you want the highest
        possible user-mode CPU utilization, but you need to watch out for increased system-mode
        CPU utilization, as it is an indication that time is being wasted on thread
        synchronization, and that you probably have too many threads. If OpenMP multi-threading
        support is built into the binary, <span class='sc'>Kelvin</span> displays the thread
        count in a  diagnostic message on startup.</li>
      <li><tt>USE_GSL</tt> Enable support for GNU Scientific Library. This is required if
        <tt>USE_OPENMP</tt> has been enabled.</li>
      <li><tt>USE_PTMALLOC</tt> and <tt>USE_HOARD</tt> Enable linking with the ptmalloc3 or
        Hoard replacement memoray allocator libraries.</li>
      <li><tt>MEMSTATUS</tt> Enables display of elapsed time and memory consumption every 30
        seconds. Disabled by default.</li>
      <li><tt>MEMGRAPH</tt> Logs elapsed time and memory consumption to a file named
        <tt>kelvin_&lt;PID&gt;_memory.dat</tt> every 30 seconds. This format can be graphed by
        <tt>gnuplot</tt> with a simple <tt>load "&lt;filename&gt;"</tt> command. Disabled by
        default.</li>
      <li><tt>POLYSTATISTICS</tt> Enables a dump of extensive polynomial build statistics at
        every 8 million raw terms and at build milestones. Disabled by default.</li>
      <li><tt>TREEEVALUATE</tt> Enables use of a polynomial evaluation technique that is optimal
        in situations where polynomials are evaluated only a few times each. This is useful only
        in rare cases, and is disabled by default.</li>
      <li><tt>FAKEEVALUATE</tt> Does not do polynomial evaluation at all. This is used for
        exercise, evaluation, and polynomial compilation. It will display a warning during
        compilation and when the program is run to the effect that evaluation is disabled and
        result file data will be meaningless. This is disabled by default, and should only be
        enabled with caution. It is documented here because an optimal performance strategy for
        analyses with extremely large amounts of evaluation is to build compiled polynomials
        without evaluation on a single-core, large memory-capacity machine, and then evaluate
        them on multi-core, smaller memory-capacity machines.</li>
      <li><tt>POLYUSE_DL</tt> Enables the use of dynamic libraries built for named polynomials
        should they be present in the current default directory. Currently the only named
        polynomials implemented in <span class='sc'>Kelvin</span> are for pedigrees, and these
        polynomials are specific to each marker set and relative trait position in the analysis.
        Compiled polynomials can be evaluated up to 10 times faster than internal polynomials,
        take up much less memory, and do not need to be rebuilt in subsequent runs. A more
        detailed discussion of polynomial compilation is currently under development. Disabled
        by default. See the section on <a href="compiledPolys.html">compiled polynomials</a>
        for details.</li>
      <li><tt>POLYCODE_DL</tt> Enables the writing of C source code implementing named
        polynomials when a previously-built dynamic library for that polynomial has not been
        found. Disabled by default, requires POLYUSE_DL. See the section on
        <a href="compiledPolys.html">compiled polynomials</a> for details.</li>
      <li><tt>POLYCOMP_DL</tt> Enables the in-run compilation of any C source code written and
        the loading and use of the dynamic library built. Disabled by default, requires
        POLYCODE_DL.  See the section on <a href="compiledPolys.html">compiled polynomials</a>
        for details.</li>
      <li><tt>POLYCHECK_DL</tt> Enables the comparison of evaluation results from any newly-built
        polynomial dynamic library with those of its source (internal) polynomial. Disabled by
        default, requires POLYCODE_DL.  See the section on <a href="compiledPolys.html">compiled
        polynomials</a> for details.</li>
      <li><tt>USE_SSD</tt> Enables the use of a file on a solid-state drive as an adjunct to
        physical memory for building particularly large likelihood polynomials.  See the section
        on <a href="SSDSupport.html">SSD support</a> for details.</li>
    </ul>
    <p>
    Once you have edited the Makefile to your satisfaction, type <kbd>make clean</kbd> and
    then <kbd>make</kbd>. This should build supporting libraries and then the two executable
    images <span class='sc'>Kelvin</span> and calc_updated_ppl.
    <p>
    Whether you use one of the prebuilt binaries or have built your own, it should be tested to
    verify functionality. Type <kbd>make test</kbd> to run a suite of tests. If you should see
    errors during the build or test, or run into other problems with
    <span class='sc'>Kelvin</span>, please send us a log of the build and test.
    <p>
    Finally, <span class='sc'>Kelvin</span> should be installed. Type <kbd>make install</kbd> to
    move it into production. 

    <h2 id='prepconfig'>Preparing The Configuration File</h2>

    <span class='sc'>Kelvin</span> typically takes all of its configuration information from a
    single file specified on the command line. This file is composed of directives that describe
    the analysis to be performed, and the locations of supporting data files. The preferred
    method for creating a configuration file is with the graphical configuration tool,
    <font color='red'>gKelvin</font>, although users may opt to edit the configuration file
    by hand. We provide a complete reference to <span class='sc'>Kelvin</span> directives
    in the <a href='Directives.html'>Directives document</a>.<p>

    In addition to the configuration file, any valid directive may be specified on the command
    line by prepending the directive with '--' (two hyphens). Any additional arguments on the
    command line will be treated as arguments to the directive, up to the end of the line, or
    the next directive, which again must be prepended with '--'. Directives on the command
    line that specify input or output files will override the values set in the configuration
    file. Directives that take a series or range of values (like
    <a href='Directives.html#traitpositions'>TraitPositions</a>), will add to the values, if any,
    specified in the configuration file. There is currently no way from the command line to
    remove values set in the configuration file.

    <p>
    <span class='sc'>Kelvin</span> supports a wide variety of analyses and options. These can
    be broken into general categories, with a small number of possibilities for each category.
    Some analyses/options are compatible with other analyses/options, some are not.

    <h3 id='twopoint_vs_multipoint'>Two-Point vs. Multipoint</h3>

    Two-point analysis is the default. Multipoint analysis is enabled with the
    <a href='Directives.html#multipoint'>Multipoint</a> directive. Multipoint analysis is
    incompatible with <a href='#ld_vs_le'>linkage disequilibrium</a> and 
    <a href='#traitmarker_vs_markermarker'>marker-to-marker</a> analyses.
  
    <h3 id='ld_vs_le'>Linkage Disequilibrium (Association) vs. Linkage Equilibrium</h3>

    Linkage equilibrium is the default for <a href='#twopoint_vs_multipoint'>
    two-point</a> analyses, and is the only option for
    <a href='#twopoint_vs_multipoint'>multipoint</a> analyses. Linkage
    Disequilibrium can be enabled for two-point analyses with the
    <a href='Directives.html#ld'>LD</a> directive.
  
    <h3 id='dt_vs_qt_vs_qtt'>Dichotomous Trait vs. Quantitative Trait vs. Quantitative Trait
    With Threshold</h3>

    A dichotomous trait model is the default. A quantitative trait model can
    be specified with the <a href='Directives.html#qt'>QT</a> directive. A quantitative trait
    model with a threshold can be specified with the <a href='Directives.html#qtt'>QTT</a>
    and <a href='Directives.html#threshold'>Threshold</a> directives. Quantitative
    trait directives are incompatible with, and the default dichotomous trait
    model is disabled for, <a href='#traitmarker_vs_markermarker'>
    marker-to-marker</a>.
  
    <h3 id='sexave_vs_sexspec'>Sex Averaged vs. Sex Specific</h3>

    By default, <span class='sc'>Kelvin</span> will perform its calculations using the sex-averaged
    centiMorgan marker positions in the map file. If sex-specific marker positions
    are available, <span class='sc'>Kelvin</span> can be made to use those with the
    <a href='Directives.html#sexspecific'>SexSpecific</a> directive. Sex-specific maps are
    not supported for <a href='#ld_vs_le'>LD</a> analyses.
  
    <h3 id='imprinting_vs_not'>Imprinting vs. Non-imprinting</h3>

    <span class='sc'>Kelvin</span> will ignore the possibility of imprinting (parent-of-origin)
    effects by default. Imprinting effects can be considered by specifying the
    <a href='Directives.html#imprinting'>Imprinting</a> directive. Imprinting analysis is
    incompatible with <a href='#traitmarker_vs_markermarker'>marker-to-marker</a>
    analyses.
  
    <h3 id='traitmarker_vs_markermarker'>Trait-to-Marker vs. Marker-to-Marker</h3>

    Trait-to-marker analysis is the default, and considers the relationship
    between a hypothetical trait locus and a marker or group of markers.
    Marker-to-marker analysis is enabled with the <a href='Directives.html#markertomarker'>
    MarkerToMarker</a> directive, and considers the relationship between
    pairs of markers only. Since no trait is considered at all under marker-to-marker,
    <a href='#dt_vs_qt_vs_qtt'>QT, QTT</a>, <a href='#imprinting_vs_not'>imprinting</a>,
    <a href='#lc_vs_nolc'>covariate-dependence</a> and <a href='#simple_vs_epistasis'>
    epistasis</a> analyses are incompatible. <a href='#twopoint_vs_multipoint'>Multipoint</a>
    analyses are also incompatible with marker-to-marker.

    <h3 id='lc_vs_nolc'>Covariate-Dependent vs. Non-Covariate-Dependent</h3>

    By default, <span class='sc'>Kelvin</span> assumes the a single penetrance vector (or
    vector of means). <span class='sc'>Kelvin</span> can be made to assign separate penetrance 
    vectors to individuals assigned to different classes, based e.g. on a covariate. Covariate
    dependence is enabled with the <a href='Directives.html#liabilityclasses'>
    LiabilityClasses</a> directive, and requires that a covariate be specified in the
    <a href='#pedfileformat'>pedigree</a> and <a href='#locusfileformat'>locus</a> files.
    Covariate dependence is incompatible with <a href='#traitmarker_vs_markermarker'>
    marker-to-marker</a> analyses, and doesn't currently work with
    <a href='#simple_vs_epistasis'>epistasis</a> analyses.

    <h3 id='simple_vs_epistasis'>Single Locus Analysis vs. Epistasis</h3>

    By default, <span class='sc'>Kelvin</span> conducts single locus analysis. The
    <a href='Directives.html#epistasis'><tt>Epistasis</tt></a> directive will provide a two-locus
    model in which penetrances may depend upon genotypes at a specified marker. Currently,
    the epistasis marker must be a SNP, and epistasis analysis doesn't work with
    <a href='#lc_vs_nolc'>covariate dependence.</a> Epistasis is incompatible with
    <a href='#traitmarker_vs_markermarker'>marker-to-marker</a> analyses.

    <h3 id='pedcounting'>Automatic Pedigree Counting</h3>

    By default, <span class='sc'>Kelvin</span> will attempt to count and bin pedigrees to
    reduce computation. It does this by examining the pedigree file prior to analysis, and
    identifying combinations of pedigree structure, phenotype and genotype that appear more
    than once. Each unique combination need only be analyzed once, thus reducing the total
    number of computations. This mechanism will realize the greatest gains for datasets
    containing mostly small pedigrees: cases-and-controls, trios and affected sib-pairs.
    <p>

    Automatic pedigree counting will be implicitly disabled if any of the following are
    true:

    <ul>
      <li>If a the <a href='Directives.html#multipoint'>MultiPoint</a> directive is present.</li>
      <li>If either the <a href='Directives.html#qt'>QT</a> or
        <a href='Directives.html#qtt'>QTT</a> directive is present.</li>
      <li>If the dataset contains microsatellites.</li>
    </ul>

    Automatic pedigree counting can be explicitly disabled with the 
    <a href='Directives.html#skippedcount'><tt>SkipPedCount</tt></a> directive.

    <h3 id='frequencyestimation'>Automatic Allele Frequency Estimation</h3>

    <span class='sc'>Kelvin</span> typically requires that allele frequencies be provided
    using the <a href='Directives.html#frequencyfile'><tt>FrequencyFile</tt></a> directive.
    However, <span class='sc'>Kelvin</span> can estimate allele frequencies in specific
    circumstances:
  
    <ul>
      <li>The analysis must be a linkage disequilibrium analysis (specified with the
        <a href='Directives.html#ld'><tt>LD</tt></a> directive).</li>
      <li>The pedigree must contain only cases-and-controls, or nuclear families (in this
        context, nuclear means exactly two founders and one or more full-sibling children
        of those founders).</li>
      <li>For any given marker, at least 30 families (any combination of cases-and-controls
        or nuclear families) must contain useful genotype information.</li>
      <li>All markers must have no more than two alleles (SNPs, typically).</li>
      <li>If the pedigree contains cases-and-controls, and the trait is dichotomous,
        then a trait prevalence must be specified with the
        <a href='Directives.html#traitprevalence'><tt>TraitPrevalence</tt></a> directive.</li>
    </ul>

    If <span class='sc'>Kelvin</span> can estimate allele frequencies, it will use those
    allele frequencies in preference to any provided using the <tt>FrequencyFile</tt>
    directive. Automatic allele frequency estimation can be explicitly disabled with the
    <a href='Directives.html#skipestimation'><tt>SkipEstimation</tt></a> directive.

    Kelvin will exit with error if allele frequency estimation is not possible and
    the <tt>FrequencyFile</tt> directive is not present.

    <h2 id='fileformats'>Data File Formats</h2>

    <span class='sc'>Kelvin</span> typically requires four input data file: a pedigree
    file, a locus file, a map file, and a frequency file. The frequency file may be 
    omitted if <span class='sc'>Kelvin</span> is to estimate allele frequencies internally.

    <h3 id='pedfileformat'>Pedigree File</h3>

    <span class='sc'>Kelvin</span> will accept either a pre- or
    post-<a href='http://linkage.rockefeller.edu/soft/linkage/sec2.7.html'>MAKEPED</a>
    format pedigree file. MAKEPED is part of the LINKAGE package. The only exception
    is that for pedigrees that contain loops, or if the proband must be explicitly 
    specified, then a post-MAKEPED pedigree file is required. While it is possible to
    create a post-MAKEPED pedigree file by hand, it's much easier to create a pre-MAKEPED
    pedigree file, and run it through MAKEPED. Each line of a pre-MAKEPED pedigree file
    must start with five columns:

    <ol>
      <li>Pedigree ID - All individuals in the same family must have the same Pedigree ID.</li>
      <li>Individual ID - The Individual ID must be unique inside a family. The combination of
        Pedigree ID and Individual ID must be unique inside the pedigree file.</li>
      <li>Father ID - this should be the Individual ID of the individual that is the father
        of the current individual. This should be '0' (zero) if the father of the current
        individual does not appear in the pedigree file.</li>
      <li>Mother ID - as above, for the current individual's mother.</li>
      <li>Sex - a <tt>1</tt> (one) if the current individual is male, or a <tt>2</tt> if
        female.</li>
    </ol>

    The remaining columns in the pedigree file are governed by the
    <a href='#locusfileformat'>locus file</a>. The pedigree file may contain any combination
    of cases-and-controls, nuclear families (defined as two founder parents and one or
    more children) or general pedigrees.
	
    <h3 id='locusfileformat'>Locus File</h3>

    This file may be referred to as a 'data file' in documentation for older versions of
    <span class='sc'>Kelvin</span>. Each line in a locus file consists of an identifier
    character, and a label. Each line corresponds to one (or two) columns in the pedigree
    file. Put another way, the locus file defines the number and order of columns in the
    pedigree file. The identifier characters, and their meanings, are as follows:

    <ul>
      <li><tt>A</tt> - Phenotype status/trait value. Corresponds to a single column in the
        pedigree file, which indicates if individuals are affected, unaffected, unphenotyped,
        or (for <a href='#dt_vs_qt_vs_qtt'>quantitative trait</a> analyses) the trait value.
        <a href='#traitmarker_vs_markermarker'>Trait-to-marker</a> analyses must have a
        phenotype status/trait line in the locus file, and the trait label will appear in
        all output files. A phenotype status/trait line need not appear for marker-to-marker
        analyses. The specific values that indicate affected, unaffected, etc., are
        controlled by the <a href='Directives.html#phenocodes'>PhenoCodes</a> configuration
        directive.</li>
      <li><tt>T</tt> - Used interchangeably with <tt>A</tt> in
        <span class='sc'>Kelvin</span>.</li>
      <li><tt>C</tt> - Covariate (Liability Class). Corresponds to a single column in the
        pedigree file. The values in this column are used to group individuals into sets with
        different trait penetrance characteristics.</li>
      <li><tt>M</tt> - Marker. Corresponds to two adjacent columns in the pedigree file. The
        values in these columns indicate the alleles of the named marker for each individual.
        Every marker that is listed in the locus file must also appear in the 
        <a href='#mapfileformat'>map file</a> and the <a href='#freqfileformat'>frequency
        file</a>.</li>
    </ul>

    <span class='sc'>Kelvin</span> currently requires that, if a phenotye status column is
    present, it must appear first, before any markers. If a covariate column is present, it
    must appear immediately after the phenotype status column, before any markers.

    <h3 id='mapfileformat'>Map File</h3>

    This file lists the positions of markers on the chromosome, in either Kosambi or
    Haldane centiMorgans. The first line of the file may optionally explicitly indicate
    the map function in use:

    <pre>mapFunction=haldane</pre>

    The rest of the file must consist of three (or more) columns, each identified by
    a column header on the first (after the optional mapFunction) line. The column headers
    are as follows:

    <ul>
      <li><tt>CHROMOSOME</tt> or <tt>CHR</tt> - The chromosome number.
        <span class='sc'>Kelvin</span> currently does not allow map files to contain markers
        from multiple chromosomes.</li>
      <li><tt>MARKER</tt> or <tt>NAME</tt> - The marker name. Marker names may contain
        numbers, letters, underscores and hyphens only.</li>
      <li><tt>POSITION</tt> or <tt>POS</tt> or <tt>KOSAMBI</tt> or <tt>HALDANE</tt> - The
        sex-averaged centiMorgan position of the marker. If the column header is 
        <tt>KOSAMBI</tt> or <tt>HALDANE</tt>, the map function will be set accordingly.</li>
      <li><tt>MALE</tt> or <tt>MALEPOSITION</tt> - The male sex-specific centiMorgan
        position of the marker. This column is only required for
	<a href='#sexave_vs_sexspec'>sex-specific</a> analyses, and will be ignored otherwise.
      <li><tt>FEMALE</tt> or <tt>FEMALEPOSITION</tt> - Same as <tt>MALE</tt>, but for
        females.</li>
      <li><tt>PHYSICAL</tt> or <tt>BASEPAIR</tt> - The basepair position of the marker.
        This column is optional. If present, this value will appear in some output files.</li>
    </ul>

    If the map function is specified by both a <tt>mapFunction</tt> line and by a column
    header, the same map function must be specified. If the map function is specified by
    neither, the default is Kosambi.

    <h3 id='freqfileformat'>Frequency File</h3>

    This file identifies the alleles for each marker, and the specifies the frequencies for
    those alleles. This file is only required if <span class=sc'>Kelvin</span> cannot 
    <a href='#frequencyestimation'>estimate allele frequencies</a> internally. The file is
    composed of a series of line groups. Each group begins with a marker line, followed by one
    or more allele lines. Similar to the <a href='#locusfileformat'>locus file</a>, each line
    starts with an identifier character:

    <ul>
      <li><tt>M</tt> - Marker line. Should be followed by the marker name only. Starts a new
        line grouop, and implicitly finishes any previous line group, if any.</li>
      <li><tt>F</tt> - Unlabeled alleles. Must be followed by only space-separted frequencies.
        This implicitly creates alleles labled with consecutive numbers, starting with '1'.
        Multiple <tt>F</tt> lines may appear in a line group; the implied labels will be
        consecutive with implied labels from previous <tt>F</tt> lines.
      <li><tt>A</tt> - Labeled alleles. Must be followed by an allele label and the
        frequency for that allele. Each allele must be specified on a separate line.</li>
    </ul>

    For any given marker, alleles must all be specified as labeled or unlabeled, although 
    different markers in the same frequency file may use either method. Note that 

    <h3 id='examplefileformat'>Example</h3>

    Pedigree File (pre-MAKEPED):
    <tt>
      <table cellspacing=2>
        <tr>
	  <td></td><td class='colno'>1</td><td class='colno'>2</td><td class='colno'>3</td>
	  <td class='colno'>4</td><td class='colno'>5</td><td class='colno'>6</td>
	  <td class='colno'>7</td><td class='colno'>8</td><td class='colno'>9</td>
	  <td class='colno'>10</td><td class='colno'>11</td><td class='colno'>12</td>
	  <td class='colno'>13</td><td class='colno'>14</td><td class='colno'>15</td>
	  <td class='colno'>16</td>
        </tr>
        <tr align='center'><td class='lineno'>1</td><td>212</td><td>101</td><td>0</td><td>0</td>
          <td>1</td><td>1</td><td>3</td><td>4</td><td>1</td><td>3</td><td>2</td><td>2</td>
          <td>1</td><td>2</td><td>1</td><td>1</td>
        </tr>
        <tr align='center'><td class='lineno'>2</td><td>212</td><td>102</td><td>0</td><td>0</td>
          <td>2</td><td>0</td><td>1</td><td>2</td><td>2</td><td>3</td><td>2</td><td>2</td>
          <td>1</td><td>1</td><td>2</td><td>2</td>
        </tr>
        <tr align='center'><td class='lineno'>3</td><td>212</td><td>103</td><td>101</td>
          <td>102</td><td>1</td><td>1</td><td>3</td><td>2</td><td>3</td><td>3</td><td>2</td>
          <td>2</td><td>1</td><td>2</td><td>2</td><td>1</td>
        </tr>
        <tr align='center'><td class='lineno'>4</td><td>212</td><td>104</td><td>0</td><td>0</td>
          <td>2</td><td>1</td><td>1</td><td>4</td><td>1</td><td>1</td><td>1</td><td>2</td>
          <td>1</td><td>2</td><td>1</td><td>1</td>
        </tr>
        <tr align='center'><td class='lineno'>5</td><td>212</td><td>105</td><td>103</td>
          <td>104</td><td>1</td><td>2</td><td>3</td><td>1</td><td>1</td><td>3</td><td>2</td>
          <td>2</td><td>2</td><td>2</td><td>2</td><td>2</td>
        </tr>
      </table>
    </tt><p>

    Locus File:
    <tt>
      <table cellspacing=2>
        <tr><td class='lineno'>1</td><td>A WebbedToes</td></tr>
        <tr><td class='lineno'>2</td><td>M rs2112</td></tr>
        <tr><td class='lineno'>3</td><td>M SNP_A-90125</td></tr>
        <tr><td class='lineno'>4</td><td>M SNP_GO-7188</td></tr>
        <tr><td class='lineno'>5</td><td>M rs8675309</td></tr>
        <tr><td class='lineno'>6</td><td>M rs1984</td></tr>
      </table>
    </tt><p>

    In this example, we have a small three-generation family. The pedigree ID (<tt>212</tt>)
    is in column 1, and individual IDs (<tt>101</tt> through <tt>105</tt>) are in column 2.
    Note that individual IDs do not need to be consecutive. The father and mother IDs are in
    columns 3 and 4, respectively. Individuals <tt>101</tt>, <tt>102</tt> and <tt>104</tt>
    are <i>founders</i> (their parents are not present in the pedigree), so their father and
    mother IDs are coded with <tt>0</tt> (zeros). Individual <tt>103</tt> is descended from
    individuals <tt>101</tt> and <tt>102</tt>; individual <tt>105</tt> is descended from 
    individuals <tt>103</tt> and <tt>105</tt>. Sex indicators are in column 5; individuals
    <tt>101</tt>, <tt>103</tt> and <tt>105</tt> are males, and individuals <tt>102</tt> and
    <tt>104</tt> are females. The remaining columns are defined by the locus file.<p>

    Line 1 of the locus file specifies a phenotype status column, which corresponds to
    column 6 in the pedigree file. This column follows the typical convention of dichotomous
    traits, coding unaffected individuals with a <tt>1</tt> (one), affected individuals with
    a <tt>2</tt> and unphenotyped individuals with a <tt>0</tt> (zero).<p>

    Each of the remaining lines in locus file specify markers, and correspond to two columns
    in the pedigree file: locus file line 2 corresponds to pedigree file columns 7 and 8, 
    line 3 to columns 9 and 10, etc. For a fun exercise, spot the impossible inheritance in
    the sample pedigree.<p>

    Installing MAKEPED is beyond the scope of this documentation, but the general command
    to process a pedigree file with MAKEPED is:

    <pre>
    $ makeped ped.pre ped.post n
    </pre>

    where <tt>ped.pre</tt> is the pre-MAKEPED file you've created, <tt>ped.post</tt> is the
    post-MAKEPED file you want to create, and the last argument is either <tt>y</tt> or
    <tt>n</tt>, depending on if any of the families in the pedigree file do or do not contain
    consanguinity loops.<p>

    Map File:
    <tt>
      <table cellspacing=2>
        <tr><td></td><td class='colno'>1</td><td class='colno'>2</td><td class='colno'>3</td></tr>
        <tr><td class='lineno'>1</td><td>CHR</td><td>MARKER</td><td>KOSAMBI</td></tr>
        <tr><td class='lineno'>2</td><td>2</td><td>rs2112</td><td>1.732</td></tr>
        <tr><td class='lineno'>3</td><td>2</td><td>SNP_A-90125</td><td>2.2361</td></tr>
        <tr><td class='lineno'>4</td><td>2</td><td>SNP_GO-7188</td><td>2.6458</td></tr>
        <tr><td class='lineno'>5</td><td>2</td><td>rs8675309</td><td>3.3166</td></tr>
        <tr><td class='lineno'>6</td><td>2</td><td>rs1984</td><td>3.6056</td></tr>
      </table>
    </tt><p>

    The map file contains the minimum three columns: chromosome (column 1), marker (column 2)
    and sex-averaged centiMorgan position (column 3). The map function is not explicitly
    specified with a <tt>mapFunction</tt> line, but the header of the column 3 specifies a
    Kosambi map. If that header was <tt>HALDANE</tt>, then a Haldance map would be specified,
    or of the header was <tt>POS</tt> or <tt>POSITION</tt>, then the map function would default
    to Kosambi.<p>

    Frequency File:
    <tt>
      <table cellspacing=2>
        <tr><td class='lineno'>1</td><td>M rs2112</td></tr>
        <tr><td class='lineno'>2</td><td>F 0.233 0.377</td></tr>
        <tr><td class='lineno'>3</td><td>F 0.300 0.090</td></tr>
        <tr><td class='lineno'>4</td><td>M SNP_A-90125</td></tr>
        <tr><td class='lineno'>5</td><td>F 0.357 0.223 0.420</td></tr>
        <tr><td class='lineno'>6</td><td>M SNP_GO-7188</td></tr>
        <tr><td class='lineno'>7</td><td>A 2 0.866</td></tr>
        <tr><td class='lineno'>8</td><td>A 1 0.134</td></tr>
        <tr><td class='lineno'>9</td><td>M rs8675309</td></tr>
        <tr><td class='lineno'>10</td><td>A 1 0.455</td></tr>
        <tr><td class='lineno'>11</td><td>A 2 0.545</td></tr>
        <tr><td class='lineno'>12</td><td>M rs1984</td></tr>
        <tr><td class='lineno'>13</td><td>F 1.000</td></tr>
      </table>
    </tt><p>

    The frequency file contains five line groups, one for each marker. Lines 1 through 3
    specify marker <tt>rs2112</tt> with four unlabeled alleles, which will be implicitly
    labeled 1, 2, 3 and 4, with corresponding frequencies of 0.233, 0.377, 0.300 and 0.090.
    Lines 4 and 5 specify marker <tt>SNP_A-90125</tt> with three unlabeled alleles. Lines 6
    through 8 specify marker <tt>SNP_GO-7188</tt> with two labeled alleles: <tt>2</tt>,
    with a frequency of 0.866, and <tt>1</tt>, with a frequency of 0.134. Lines 9 through 11
    specifiy marker <tt>rs8675309</tt> with two labeled alleles. Lines 12 and 13 specify
    a marker <tt>rs1984</tt> with a single unlabeled allele with a frequency of 1.000. Note
    than in no case are labeled and unlabelled alleles mixed within a single marker.

    <h2 id='running'>Running The Programs</h2>

    Once you have installed <span class='sc'>Kelvin</span>, you can run it from your data
    directory, where you keep your configuration and data files. <span class='sc'>Kelvin</span>
    takes only one parameter, which is the name of the configuration file, e.g.:

    <pre>
    _$ kelvin kelvin.conf
    </pre>

    Remember that if you did not specify absolute paths for output files in the configuration
    file, they will be written relative to your current directory.
    <p>
    It is often important to capture all output from a run into a file so that you may review
    it more after the run completes, or send it to us for diagnosis. The following command
    (using <tt>sh</tt>/<tt>ksh</tt>/<tt>bash</tt> syntax) runs <span class='sc'>Kelvin</span>
    with all output redirected to a file called <tt>kelvin.out</tt>:

    <pre>
    _$ kelvin kelvin.conf > kelvin.out 2>&1
    </pre>

    Or, using <tt>csh</tt>/<tt>tcsh</tt> syntax:

    <pre>
    _$ kelvin kelvin.conf >& kelvin.out
    </pre>

    If you do need to send us information for diagnosis, please include the configuration and
    data files along with the output from the run.

    <h3>Status Messages</h3>

    When <span class='sc'>Kelvin</span> is run, it first displays version, build and
    configuration information. All messages are prefaced with the current date and time. Messages
    fall into three categories:
    <ol>
      <li><b>Progress messages</b>, which begin with an at-sign (@) and elapsed time from the
        start of the run. They are indented according to their level of detail. They
	are relatively inconsequential, and intended only to let you know how much progress
        has been made on the analysis. You can control how much detail is provided and how
        frequently messages are displayed with the
        <a href='Directives.html#progresslevel'>ProgressLevel</a> and
        <a href='Directives.html#progressdelayseconds'>ProgressDelaySeconds</a> directives. </li>
      <li><b>Diagnostic messages</b>, which begin with a severity indication and then provide a description
	of the difficulty encountered. Severity levels are <b>WARNING</b>, which indicates a potential problem
	that requires your attention in order to ensure the integrity of the analysis results; 
	<b>ERROR</b>, which indicates
	some kind of analysis problem that will prevent the run from continuing; and <b>FATAL</b>, which indicates
	an internal integrity failure in <span class='sc'>Kelvin</span> that should be brought to the attention of
	the developers. </li>
	<li><b>Informational messages</b>, are essentially everything else, and while helpful, do not necessarily
	  need to be reviewed in order to ensure the integrity of the analysis results.
</ol>

    For example:
<p>
  <table cellspacing=2>
    <tr><td class='lineno'><tt> 1 </tt></td>
<td><tt>10/01/15 08:26:25 @0s, Initializing</tt></td></tr>
    <tr><td class='lineno'><tt> 2 </tt></td>
<td><tt>10/01/15 08:26:25 kelvin V0.38.1 edit 2463 built Jan 14 2010 15:04:43</tt></td></tr>
    <tr><td class='lineno'><tt> 3 </tt></td>
<td><tt>10/01/15 08:26:25 Compiler 4.1.1 20070105 (Red Hat 4.1.1-52)</tt></td></tr>
    <tr><td class='lineno'><tt> 4 </tt></td>
<td><tt>10/01/15 08:26:25 Using GNU Scientific Library (GSL) statistical functions instead of internal ones</tt></td></tr>
    <tr><td class='lineno'><tt> 5 </tt></td>
<td><tt>10/01/15 08:26:25 OpenMP-enabled w/maximum of 6 thread(s).</tt></td></tr>
    <tr><td class='lineno'><tt> 6 </tt></td>
<td><tt>10/01/15 08:26:25 GCC optimization level 2 enabled</tt></td></tr>
    <tr><td class='lineno'><tt> 7 </tt></td>
<td><tt>10/01/15 08:26:25 To check status (at some risk), type CTRL-\ or type "kill -3 22860"</tt></td></tr>
    <tr><td class='lineno'><tt> 8 </tt></td>
<td><tt>10/01/15 08:26:25 Using configuration file kelvin.conf</tt></td></tr>
    <tr><td class='lineno'><tt> 9 </tt></td>
<td><tt>10/01/15 08:26:25 Computation is done in polynomial mode</tt></td></tr>
    <tr><td class='lineno'><tt> 10 </tt></td>
<td><tt>10/01/15 08:26:25 polynomialScale is 1 (1-10, 1 is default)</tt></td></tr>
    <tr><td class='lineno'><tt> 11 </tt></td>
<td><tt>10/01/15 08:26:25 Integration is done numerically (dkelvin)</tt></td></tr>
    <tr><td class='lineno'><tt> 12 </tt></td>
<td><tt>10/01/15 08:26:25 Further progress will be displayed at 90 second intervals</tt></td></tr>
    <tr><td class='lineno'><tt> 13 </tt></td>
<td><tt>10/01/15 08:26:25 6TL of 20AL*6GF*275pv(1LC) space for 24 pedigree(s)</tt></td></tr>
    <tr><td class='lineno'><tt> 14 </tt></td>
<td><tt>10/01/15 08:26:25 Trait-to-marker, Sex-Averaged Multipoint (w/4 loci), Dichotomous Trait.</tt></td></tr>
    <tr><td class='lineno'><tt> 15 </tt></td>
<td><tt>10/01/15 08:26:25 @0s, Performing analysis</tt></td></tr>
    <tr><td class='lineno'><tt> 16 </tt></td>
<td><tt>10/01/15 08:26:26&nbsp;&nbsp;&nbsp;&nbsp;@1s, Starting with position 1 of 6</tt></td></tr>
    <tr><td class='lineno'><tt> 17 </tt></td>
<td><tt>10/01/15 08:27:56&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@1m30s, Building polynomial w/pedigree: 12 (12/24)</tt></td></tr>
    <tr><td class='lineno'><tt> 18 </tt></td>
<td><tt>10/01/15 09:26:20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@2m32s, Building polynomial w/pedigree: 24 (24/24)</tt></td></tr>
    <tr><td class='lineno'><tt> 19 </tt></td>
<td><tt>10/01/15 09:26:20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@3m59s, Evaluating polynomials, currently at 8.1e+03 iterations</tt></td></tr>
  </table>
<p>
  Lines 1 and 15-19 are progress messages.
  Line 2 is the <span class='sc'>Kelvin</span> major version and build information.
  Next is the compiler verison number (line 3), followed by important compilation conditionals as
  specified in the Makefile and run characteristics as influenced by environment variables. In
  the example given, OpenMP support status and the number of threads used (line 5)
  is displayed. Other options and their
  corresponding messages are covered in the <a href="#building"> Building and Testing</a>
  section.
    <p>
      Following this is a line (7) describing the action to take in order to force an early display
      of progress messages.
You can use this as a sort of "pulse check" to make sure <span class='sc'>Kelvin</span> is still alive
and well and making progress should there be a lull between automatic progress updates. If
you are running <span class='sc'>Kelvin</span> interactively, you can 
perform this "pulse check" by type <tt>CTRL-\</tt> (that means to hold down the CTRL
or CONTROL key while pressing the backslash). If you are running under cygwin, you will first need to
type <tt>stty quit ^C</tt> to make this to work. Note that the aforementioned <tt>^C</tt> is actually the two
character sequence of 'carat' (shifted 6) and 'C'. Pressing <tt>CTRL-\</tt> sends a SIGQUIT signal to <span class='sc'>Kelvin</span>, which it
interprets as a request for status. If you are running <span class='sc'>Kelvin</span> as a detached
process or in a batch queue, you can send a SIGQUIT to the process by being logged-into the same
node as it is running on, and using the <tt>kill</tt> command
as described in the diagnostic output. Note that the signal number (-3 in the example) can be
different from platform-to-platform, and the
process ID (26577 in the example) will be different from run-to-run. The "at some risk" bit is because
some status information is displayed asynchronously, i.e. regardless of the current context of the
evaluation, and has been known, albeit extremely infrequently, to crash the program.
<p>
Next, location of the configuration file (line 8) and the analysis and run characteristics as determined from that file
(lines 9-12) are displayed. The last of these (lines 13 and 14) are a terse description of the scope and nature
of the analysis, for example a fixed-grid run might display:
<pre>
6TL of 20AL*6GF*275pv(1LC) space for 24 pedigree(s)
Trait-to-marker, Sex-Averaged Multipoint (w/4 loci), Dichotomous Trait.
</pre>
This indicates that for each of 6 trait loci, PPLs will be computed using the closest 4 markers for the 
standard 3-dimensional trait model space of 20 alpha values by 6 gene frequencies and 275 penetrance 
values (from 1 liability class). A two-point analysis will provide slightly different information, e.g.:
<pre>
9 pair(s)*51Th of 20AL*6GF*275pv(1LC) space for 11 pedigree(s)
Trait-to-marker Two-Point, Dichotomous Trait, Disequilibrium.
</pre>
This indicates that for each of the 9
pairs of loci, 51 values of theta will be evaluated in the standard 3-dimensional trait model space of 20 alpha 
values by 6 gene 
frequencies and 275 penetrance values (from 1 liability class), which is 459 integrations over a 33,000 point grid. 
Likelihoods for the evaluation  will be derived from the phenotypic and genotypic information in 11 pedigrees.
This can give you an idea of how complex your run is.
<p>
Note that dynamic grid runs cannot anticipate the number of sample points that will be used in the trait 
model space, and therefore will display something like:
<pre>
6TL over a dynamic space for 24 pedigree(s)
Trait-to-marker, Sex-Averaged Multipoint (w/4 loci), Dichotomous Trait.
</pre>
<p>
Finally, progress indicators are displayed up through the end of the run.

    <h2 id='outputfiles'>Contents Of Output Files</h2>

    <span class='sc'>Kelvin</span> produces several different files of results depending upon the
    type of analysis performed.

    <h3>Bayes Ratio File</h3>

    Integrated likelihood ratio (BR) information is produced for all runs. It is written to the
    file <tt>br.out</tt> by default, although that filename is configurable. The first line
    written to the file is a comment that contains the <span class='sc'>Kelvin</span> version
    number, e.g.:

    <pre>
    # Version V0.38.1 edit 2463
    </pre>

    This is included so that results can be associated with the version of
    <span class='sc'>Kelvin</span> that produced them, and will allow for variations in
    algorithms, file formats and data precision. Subsequent lines contain different information
    for each type of analysis being performed:

    <h4 id='br-multipoint'>Multipoint Analysis</h4>

    For multipoint runs, a single table with a header line is output. The table consists of one
    row for each trait position. Columns are whitespace-delimited:

    <ol>
      <li>Chr - chromosome number</li>
      <li>Position - trait positions in centiMorgans</li>
      <li>PPL - the posterior probability of linkage. This is rounded to two places if .025 or
        greater, and three decimal places otherwise.</li>
      <li>BayesRatio - integrated likelihood ratio.</li>
      <li>MarkerList(...) - parenthesised list of the closest N markers for each position, where
        N is the number of markers being considered at a time. Header text reflects actual count
	of columns.</li>
    </ol>

    <h4 id='br-twopoint'>Two-point Analysis</h4>

    Two-point analyses output separate tables for each locus. Each table is prefaced with a
    comment line that details the chromosome, name and position of the current marker, e.g.:

    <pre>
    # Seq: 1 Chr: 1 Trait: disease Marker: 1_3_mrk Position: 0.9900
    </pre>

    and consists of one row of whitespace-delimited values for each value of D' (in the case of an
    LD analysis) and theta. Columns are whitespace-delimited:

    <ol>
      <li>multiple columns of D<i>nm</i> - the D' (d-prime) between allele <i>m</i> of the trait
        and allele <i>n</i> of the marker. There will be (<i>n</i>-1)*(<i>m</i>-1) columns
        present for linkage disequilibrium runs. These columns are not present on LE runs because
        LE is a special case of LD where D' is always 0. Note that the number of D' columns will
        vary within an analysis when the number of marker alleles varies.</li>
      <li>Theta(M,F) - male and female theta values. They are identical for a sex-averaged
        analysis.</li>
      <li>BayesRatio - integrated likelihood ratio.</li>
    </ol>

    <h3 id='ppl-twopoint'>Posterior Probability of Linkage (PPL) File</h3>

    Separate PPL information is produced for two-point and marker-to-marker analyses. It is
    written to the file <tt>ppl.out</tt> by default, although that filename in configurable.
    The first line is a comment that contains the <span class='sc'>Kelvin</span> version number
    as with the Bayes Ratio file. The next line is a header, followed by one line of
    whitespace-delimited values for each marker in the run.

    <ol>
      <li>Chr - chromosome number.</li>
      <li>Trait/Marker1 - trait name for trait-to-marker analysis, first marker name for
        marker-to-marker analysis.</li>
      <li>Position1 - first marker position in centiMorgans (not present for trait-to-marker
        analysis).</li>
      <li>Marker/Marker2 - marker name for trait-to-marker analysis, second marker name for
        marker-to-marker analysis.</li>
      <li>Position/Position2 - marker position in centiMorgans.</li>
      <li>PPL - posterior probability of linkage. Evidence for or against linkage to the disease 
        locus, not allowing for disequilibrium, i.e., D' is fixed at 0. It is rounded to two
        places if .025 or greater, and three decimal places otherwise.</li>
      <li>PPL(LD) (only for LD analyses) - evidence for or against linkage allowing for
        disequilibrium. The PPL(LD) is rounded to two places if .025 or greater, and three
        decimal places otherwise.</li>
      <li>PPLD|L (only for LD analyses) - Posterior probability of LD, given linkage. The
        PPLD|L is rounded like the PPL(LD).</li>
      <li>PPLD(L) (only for LD analyses) - posterior probability of linkage andlinkage
        disequilibrium, The PPLD(L) is rounded like the PPL(LD).</li>
    </ol>

    <h3 id='modfile'>Maximizing Model (MOD) File</h3>

    All analyses can optionally write maximizing models to a separate file, named
    <tt>mod.out</tt> by default, although that name is configurable. In all cases, the
    first line is a comment line that identifies the version of <span class='sc'>Kelvin</span>
    that created the file.

    <h4>Multipoint Analysis</h4>

    For multipoint runs, a single table with a header line is output. The table consits of
    one row for each trait position, containing the values of all trait parameters that
    maximized the likelihood at that position. Columns are whitespace delimited:

    <ol>
      <li>Chr - chromosome number</li>
      <li>Position - trait positions in centiMorgans</li>
      <li>MOD - maximum LOD score achieved at this position.</li>
      <li>Alpha - the value of alpha that maximized the HLOD score.</li>
      <li>DGF - the maximizing disease gene frequency.</li>
      <li>LC<tt>n</tt>PV(...) - the maximizing penetrance vector, one for each liability
         (covariate) class in the analysis. For dichotomous trait runs, it is three (or four,
         if imprinting effects are being considered) columns of the maximizing penetrance for
         DD, Dd, (dD,) and dd. For quantitative trait (QT) runs with the normal distribution,
         it is three (or four) columns of means followed by three (or four) of standard
         deviations for the maximizing distributions for DD, Dd, (dD,) and dd, followed by the
         threshold in the case of the QTT model. Quantitative trait runs with the Chi-Square
         distribution have only three (or four) columns of degrees of freedom followed by the
         threshold. Values are comma-separated and enclosed in parentheses. Header text
         reflects actual count and nature of columns.</li>
    </ol>

    <h4>Two-point Analysis</h4>

    For two-point analyses, a separate table is output for each marker, or pair of markers,
    in the case of marker-to-marker analyses. Each table is prefaced with a comment line that
    details the chromosome, name and position of the marker (or markers). For marker-to-trait
    analyses, this line is identical to that in the Bayes Ratio file. For marker-to-marker, the
    line looks like:

    <pre>
    # Seq: 2 Chr 1: Marker1: 1_3_mrk Position1: 0.9900 Marker2: 1_4_mrk Position2: 1.3200
    </pre>

    Each table is prefaced with a header line. Columns are whitespace-delimited:

    <ol>
      <li>Case - Identifies the model as maximizing overall, or maximizing where theta is 0, or
        (in the case of LD analyses) as maximizing where D' is 0.</li>
      <li>MOD - maximum LOD score achieved for the current Case.</li>
      <li>multiple columns of D' values, as described for two-point Bayes Ratio files. For
        LD analyses only.</li>
      <li>Theta(M,F) - as described for two-point Bayes Ratio files.</li>
      <li>Alpha, DGF and LC<tt>n</tt>PV(...) - as described for multipoint maximizing mode
        files. Not present in marker-to-marker analyses.</li>
      <li>R<sup>2</sup> - the standard LD correlation coefficient. Only present in 
        marker-to-marker runs.</li>
    </ol>

    <h2 id='sequpdate'>Sequential Updating</h2>
    Sequential updating is a method for combining the results of multiple analyses in
    post-processing. Sequential updating tools are included in the <span class='sc'>Kelvin</span>
    distribution.

    <h3 id='calcupdatedppl'>calc_updated_ppl</h3>

    This program takes one or more <span class='sc'>Kelvin</span>-format Bayes ratio files,
    and produces PPL (and, if appropriate) PPLD statistics. If a single Bayes ratio file is
    provided, <tt>calc_updated_ppl</tt> simply reproduces the PPL statistics generated by the 
    original analysis. If multiple Bayes ratio files are provided, the Bayes ratios are
    sequentially updated and new PPL statistics are generated.<p>

    <tt>calc_updated_ppl</tt> can also use the PPL statistics generated by a multipoint analysis
    as prior probabilities when calculating the PPLD(L) (posterior probability of linkage and LD)
    statistic. In this way, the results of a linkage analysis can be combined with an LD
    analysis. Both the linkage and LD analyses should have been conducted using the same map.<p>

    By default, <tt>calc_updated_ppl</tt> expects Bayes ratio files from two-point, sex-averaged
    analyses, in the most recent format, and displays PPL statisics. Bayes ratio files created
    using older versions of <span class='sc'>Kelvin</span> must be
    converted to the current format using the provided <a href='#convertbr'>conversion script</a>.
    All input files must contain Bayes ratios for the same markers, in the same order. The
    default behavior can be modified with a number of command line flags:

    <div class='indent' cellpadding='2'>
      <table>
	<tr>
          <td valign='top'><tt>-m</tt>,&nbsp<tt>--multipoint</tt></td>
          <td>Specifies that the input Bayes ratio files are from multipoint analyses. PPLs
            will be calculated at each position that appears in any of the input files. If a
            given position does not appear in all files, the updated PPL will be calculated 
            using only the files in which the position does appear.</td>
        </tr>
	<tr>
          <td valign='top'><tt>-s</tt>,&nbsp<tt>--sexspecific</tt></td>
	  <td>Specifies that the input Bayes ratio files were created using a sex-specific
             map.</td>
        </tr>
	<tr>
          <td valign='top'><tt>-M&nbsp<i>mapfile</i></tt>, <tt>--mapin&nbsp<i>mapfile</i></tt></td>
	  <td>Specifies a map file for input. Only valid with two-point Bayes ratio files.
            When a map is provided, individual Bayes ratio files are permitted to omit some
            markers. Each file may only contain markers that appear in the map, and they must
            all appear in the same order as they appear in the map. Updated statistics will be
            calculated for each marker that appears in at least one Bayes ratio file.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-f</tt>,&nbsp<tt>--forcemap</tt></td>
          <td>Specifies that marker position information in Bayes ratio files should be
            ignored, and marker positions from a supplied map should be used instead. Requires
            that a map have been provided with <tt>--mapin</tt>.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-R&nbsp<i>pplfile</i></tt>, <tt>--pplout&nbsp<i>pplfile</i></tt></td>
          <td>Specifies the name of a file to which updated PPL statistics should be written,
            rather than displaying them.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-O&nbsp<i>brfile</i></tt>, <tt>--partout&nbsp<i>brfile</i></tt></td>
          <td>Specifies the name of a file to which updated Bayes ratios should be written.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-P&nbsp<i>pplfile</i></tt>, <tt>--pplin&nbsp<i>pplfile</i></tt></td>
          <td>Specifies the name of a file containing multipoint PPLs which should be used
            as prior probabilities for calculating the PPLD(L) statistic. This new statistic is
            labled cPPLD (combined PPLD). For each marker in the LD analysis, the prior
            probability will be sampled from the multipoint PPLs. If there is no PPL at the
            exact marker position, one will be estimated using linear interpolation.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-r</tt>,&nbsp<tt>--relax</tt></td>
          <td>Specifies that marker names in Bayes ratio files should be ignored, and that updated
            statistics should be calculated using Bayes ratios from all files in the order in
            which it appears. This option should be used WITH CAUTION, since it implies that
            the same markers appear in different analyses with different names, which is probably
            not the case.</td>
        </tr>
        <tr>
          <td valign='top'><tt>-?</tt>,&nbsp<tt>--help</tt></td>
          <td>Displays usage help.</td>
        </tr>
      </table>
    </div>

    <h4>Examples</h4>

    Here's a simple example of updating across two two-point Bayes ratio files, and saving the
    updated statistics to a file:

    <pre>
_$ calc_updated_ppl br1.out br2.out > updated-ppl.out
    </pre>

    Here's an example of updating across three multipoint Bayes ratio files, and requests that
    the updated Bayes Ratios be written to the file <tt>br-updated.out</tt>:

    <pre>
_$ calc_updated_ppl -m --partout br-updated.out br1.out br2.out > updated-ppl.out
    </pre>

    Here's an example of updating across three two-point Bayes ratio files, each of which 
    contains a slightly different subset of the total set of markers. A map file is required
    in this case:

    <pre>
_$ calc_updated_ppl --mapin complete.map br1.out br2.out br3.out > updated-ppl.out
    </pre>

    Here's an example of using multipoint PPLs with a single two-point, LD Bayes ratio
    file. This just recalculates the PPLD statistics for the given Bayes ratios, but uses
    the multipoint PPLs as priors for the PPLD(L) statistic (producing the cPPLD statistic):

    <pre>
_$ calc_updated_ppl --pplin mp-ppl.out br.out > cppld.out
    </pre>

    <h3 id='convertbr'>convert_br.pl</h3>

    This Perl script will convert <tt>avghet.out</tt> or <tt>br.out</tt> files generated by
    older versions of <span class='sc'>Kelvin</span> into the most recent format. Output is
    always to the terminal.<p>

    <div class='indent' cellpadding='2'>
      <table>
	<tr>
          <td valign='top'><tt>-m&nbsp<i>mapfile</i></tt></td>
          <td>If the Bayes ratio file was generated by <span class='sc'>Kelvin</span>
            0.36.0 or earlier, it will contain no Position column. In this case, a mapfile
            may be provided, and the marker positions from the mapfile will be used to 
            fill position information in the output Bayes ratio file. If the string
	    <tt>nomap</tt> is passed, instead of a mapfile, then the marker positions
            will be filled with sequential integers.</td>
        </tr>
	<tr>
          <td valign='top'><tt>-c&nbsp<i>chrnum</i></tt></td>
	  <td>If the Bayes ratio file was generated by <span class='sc'>Kelvin</span>
          0.34.2 or earlier, there will be no Chromosome column. In this case, the
          user must specify the chromsome.</td>
        </tr>
      </table>
    </div>

    <h4>Examples</h4>

    Here's an example that captures the output to a file:

    <pre>
_$ convert_br.pl -c 12 -m mapfile.dat avghet-old.out > br-new.out
    </pre>

    Note in this example, <tt>-c 12</tt> specifies that <tt>avghet-old.out</tt> contains
    data from an analysis of chromosome 12, and that marker position information is in
    the file <tt>mapfile.dat</tt>. A full description of command line syntax is available by
    typing:

    <pre>
_$ convert_br.pl --help
    </pre>

    <hr>
    $Date$
  </body>

</html>
