<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head><meta content="text/html; charset=ISO-8859-1" http-equiv="content-type"><title>graphApp manual</title></head>
<body><h2>Outline</h2><hr style="width: 100%; height: 2px;"><br><br><a href="#Running_graphApp">Running graphApp</a><br><a href="#graphApp">graphApp</a><br><a href="#Toolbar_buttons">Toolbar buttons</a><br>&nbsp;&nbsp;&nbsp; <a href="#open_graph_file">Open graph file</a><br>&nbsp;&nbsp;&nbsp; <a href="#import_data_from_text_file">Import data from text file</a><br>&nbsp;&nbsp;&nbsp; <a href="#save_graph_file">Save graph file</a><br>&nbsp;&nbsp;&nbsp; <a href="#export_to_image">Export to image</a><br>&nbsp;&nbsp;&nbsp; <a href="#print">Print</a><br>&nbsp;&nbsp;&nbsp; <a href="#insert_horizontal_line">Insert horizontal line</a><br>&nbsp;&nbsp;&nbsp; <a href="#zoom_to_chr">Zoom to Chromosome</a><br>&nbsp;&nbsp;&nbsp; <a href="#home">Home</a><br>&nbsp;&nbsp;&nbsp; <a href="#back">Back</a><br>&nbsp;&nbsp;&nbsp; <a href="#forward">Forward</a><br>&nbsp;&nbsp;&nbsp; <a href="#pan">Pan</a><br>&nbsp;&nbsp;&nbsp; <a href="#zoom_rectangle">Zoom Rectangle</a><br>&nbsp;&nbsp;&nbsp; <a href="#zoom_x">Zoom X-Axis</a><br>&nbsp;&nbsp;&nbsp; <a href="#zoom_y">Zoom Y-Axis</a><br>&nbsp;&nbsp;&nbsp; <a href="#annotate_point">Annotate Point</a><br>&nbsp;&nbsp;&nbsp; <a href="#annotate_threshold">Annotate Threshold</a><br>&nbsp;&nbsp;&nbsp; <a href="#annotate_peak">Annotate Peak</a><br>&nbsp;&nbsp;&nbsp; <a href="#annotate_valley">Annotate Valley</a><br>&nbsp;&nbsp;&nbsp; <a href="#move_annotation">Move Annotation</a><br>&nbsp;&nbsp;&nbsp; <a href="#configure_plot">Configure Plot</a><br><a href="#File_Menu_Bar">File Menu Bar</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_open_graph_file">Open Graph File</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_save_graph_file">Save Graph File</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_save_as_graph_file">Save As Graph File</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_import_data_from_file">Import Data from File</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_import_data_from_remote_file">Import Data from Remote File</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_export_graph_to_color_image">Export Graph to Color Image</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_export_graph_to_grayscale_iamge">Export Graph to Grayscale Image</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_page_setup">Page Setup</a><br>&nbsp;&nbsp;&nbsp; <a href="#fmb_print">Print</a><br><a href="#Edit_Menu_Bar">Edit Menu Bar</a><br>&nbsp;&nbsp;&nbsp; <a href="#emb_clear_graph">Clear Graph</a><br>&nbsp;&nbsp;&nbsp; <a href="#emb_copy_graph">Copy Graph</a><br>&nbsp;&nbsp;&nbsp; <a href="#emb_remove_chr">Remove Chromosomes</a><br><a href="#Options_Menu_Bar">Options Menu Bar</a><br>&nbsp;&nbsp;&nbsp; <a href="#omb_configure_graph">Configure Graph</a><br>&nbsp; &nbsp; <a href="#omb_configure_linesets">Configure Line Sets</a><br>&nbsp; &nbsp; <a href="#omb_configure_all_lines">Configure All Lines</a><br>&nbsp; &nbsp; <a href="#omb_configure_genemarkers">Configure Gene Markers</a><br>&nbsp;&nbsp;&nbsp; <a href="#omb_set_view_limits">Set x-axis, y-axis limits</a><br><a href="#Insert_Menu_Bar">Insert Menu Bar</a><br>&nbsp;&nbsp;&nbsp; <a href="#imb_insert_hline">Insert Horizontal Line</a><br>&nbsp; &nbsp; <a href="#imb_insert_genemarkers">Insert Gene Markers</a><br><a href="#Overlap_Menu_Bar">Overlap Menu Bar</a><br>&nbsp;&nbsp;&nbsp; <a href="#omb_insert_overlap_chr_from_file">Insert Overlapping Chromosome from File</a><br>&nbsp;&nbsp;&nbsp; <a href="#omb_overlap_existing_chr">Overlap existing chromosomes</a><br><a href="#Annotations">Annotations</a><br>&nbsp;&nbsp;&nbsp; <a href="#ann_add_annotation">Add Annotation</a><br>&nbsp;&nbsp;&nbsp; <a href="#ann_remove_annotation">Remove Annotation</a><br>&nbsp;&nbsp;&nbsp; <a href="#ann_edit_annotations">Editing Annotations</a><br>&nbsp;&nbsp;&nbsp; <a href="#ann_annotation_layout">Annotation Layout</a><br><a href="#command_line">Command Line Options</a><br><a href="#Mathematical_Expressions">Mathematical Expressions</a><br><h2><a name="Running_graphApp"></a></h2><h2>Running graphApp</h2><hr style="width: 100%; height: 2px;"><br><br>There are two ways to run graphApp:<br><br>1. <span style="font-weight: bold;">Run from the Cluster</span>
- Log onto the cluster with an X11 session and type 'graphApp' from any
directory.&nbsp;&nbsp; This will cause the program to be run from the
cluster.&nbsp; It will use Linux icons and menus.&nbsp; The program
speed and responsiveness may be a little slow when running from the
cluster.<br><br>When using graphApp from the cluster, you may notice
that the mouse turn yellow, or just completely disappears.&nbsp; This
problem is from a bug in Mac OSX.&nbsp; To fix this, log into the
cluster and type:<br><br><span style="font-family: monospace;">mkdir ~/.icons</span><br style="font-family: monospace;"><span style="font-family: monospace;">cd ~/.icons</span><br style="font-family: monospace;"><span style="font-family: monospace;">cp -R /usr/share/icons/HighContrast .</span><br style="font-family: monospace;"><span style="font-family: monospace;">ln -s HighContrast default</span><br><br><br>2. <span style="font-weight: bold;">Run from local machine</span>
- Copy the source files and run graphApp from a local computer.&nbsp;
The source files are located on the cluster at
/usr/local/share/graphApp.&nbsp; Source files are also located on the
file server, under the software folder.&nbsp; You will also need the
following prerequisites in order to run graphApp from your computer.<br><br>&nbsp;&nbsp; &nbsp;- Python 2.5<br>&nbsp;&nbsp; &nbsp;- wxPython<br>&nbsp;&nbsp; &nbsp;- Matplotlib 0.90.1<br>&nbsp;&nbsp; &nbsp;- NumPy<br>&nbsp;&nbsp; &nbsp;- Python Imaging Library (PIL)<br>&nbsp; &nbsp; - PyXML &nbsp; &nbsp;<br><br>If you are trying to install these packages on Mac OSX, all of them can be found <a target="_blank" href="http://pythonmac.org/packages/py25-fat/index.html">here</a>, except for PyXML. You will have to go to their homepage, <a href="http://pyxml.sourceforge.net/">here</a> to download. &nbsp;Once downloaded, follow their instructions to install the module.<br><br>Once
the source files are copied and the prerequisites are installed, simply
open a command prompt, goto the directory where the sources files are
located, and type <br><br>python graphApp.py<br><br><h2><a name="graphApp"></a></h2><h2>graphApp</h2><hr style="width: 100%; height: 2px;"><br>Here is a screenshot diagram of graphApp (Linux version).<br><br><img style="width: 854px; height: 659px;" alt="" src="graphApp_manual_files/screenshot.png"><br><br><span style="font-weight: bold;">comment box</span> - a text box for placing comments about the graph<br><br><span style="font-weight: bold;">info panel</span> - a panel listing all previously annotated peaks and valleys that are currently visible in the graph<br><br><span style="font-weight: bold;">sidebar anchor</span> - drag from this notched area to resize the comment box and info panel to give more or less to the graph<br><br><h2><a name="Toolbar_buttons"></a></h2><h2>Toolbar buttons</h2><hr style="width: 100%; height: 2px;"><br><span style="font-weight: bold;">Tip</span>: If you hover over a toolbar button for awhile, a tooltip of what it does appears.<br><br><a name="open_graph_file"></a><br><img style="width: 47px; height: 42px;" alt="" src="graphApp_manual_files/open_graph.png">&nbsp;&nbsp; <span style="font-weight: bold;">Open grx file</span><br>Opens
a .grx file. A .grx file is a special file format used by graphApp to
save the state of a graph, which can later be opened by graphApp.<br><br><a name="import_data_from_text_file"></a><br><img style="width: 47px; height: 42px;" alt="" src="graphApp_manual_files/import_text.png">&nbsp;&nbsp; <span style="font-weight: bold;">Import data from text file</span><br>Opens a text file containing data (See 'Import Data from File').<br><br><a name="save_graph_file"></a><br><img style="width: 47px; height: 42px;" alt="" src="graphApp_manual_files/save_graph.png">&nbsp;&nbsp; <span style="font-weight: bold;">Save grx file</span><br>Saves the currently shown graph as a .grx file.<br><br><a name="export_to_image"></a><br><img style="width: 47px; height: 42px;" alt="" src="graphApp_manual_files/export_image.png">&nbsp;&nbsp; <span style="font-weight: bold;">Export to image </span><br>Saves
the currently shown graph as an image file.&nbsp; Note that if no
extension is given, the default image format is png.&nbsp; Other
formats are not recommended.<br><br><a name="print"></a><br><img style="width: 37px; height: 34px;" alt="" src="graphApp_manual_files/print.png">&nbsp;&nbsp; <span style="font-weight: bold;">Print</span> <br>Print the current graph.<br><br><span style="font-weight: bold;"><a name="insert_horizontal_line"></a><br><img style="width: 38px; height: 29px;" alt="" src="graphApp_manual_files/horizontal_line.png">&nbsp;&nbsp; Insert horizontal line</span><br>Use
this button to insert a black horizontal line that runs through the
entire graph.&nbsp; When this button is pressed, a dialog box will
appear asking what y-values to place the line(s).&nbsp; If multiple
horizontal lines are wanted, separate values with commas (e.g. '0.1,
0.2, 0.3').<br><br><span style="font-weight: bold;"><a name="zoom_to_chr"></a><br><img style="width: 136px; height: 41px;" alt="" src="graphApp_manual_files/zoom_to_chr.png">&nbsp;&nbsp; Zoom to Chromosome</span><br>These
two widgets allow you to zoom in on one chromosome.&nbsp; Use the
drop-down box on the left to select which chromosome to zoom in on,
then press the magnifying glass icon on the right to force the graph to
only show that chromosome.<br><br><a name="home"></a><br><img style="width: 36px; height: 31px;" alt="" src="graphApp_manual_files/home.png">&nbsp;&nbsp; <span style="font-weight: bold;">Home</span><br>Clicking
on this button will show the base view.&nbsp; This usually means the
x-axis shows all chromosomes in the graph, and the y-axis is set to 0
and 1.<br><br><a name="back"></a><br><img style="width: 35px; height: 32px;" alt="" src="graphApp_manual_files/back.png">&nbsp;&nbsp; <span style="font-weight: bold;">Back</span><br>Go to the previous view.&nbsp; A new view is set when a zoom or a pan occurs.<br><br><a name="forward"></a><br><img style="width: 38px; height: 31px;" alt="" src="graphApp_manual_files/forward.png">&nbsp;&nbsp; <span style="font-weight: bold;">Forward</span><br>Go to the next view.&nbsp; The opposite effect of the Back button.<br><br><a name="pan"></a><br><img style="width: 37px; height: 31px;" alt="" src="graphApp_manual_files/pan.png">&nbsp;&nbsp; <span style="font-weight: bold;">Pan</span><br>Press
once to set the graph to pan mode.&nbsp; Press again to get out of pan
mode.&nbsp; While in pan mode, you can drag the graph using the left
mouse button.&nbsp; There are two special options pertaining to
this.&nbsp; If you hold the 'x' button on the keyboard while dragging
with the left mouse button, the graph will only move in the x
direction.&nbsp; If you hold the 'y' button on the keyboard while
dragging, the graph will only move in the y direction.&nbsp; Also in
pan mode you can drag with the right mouse button.&nbsp; Dragging with
the right mouse button will case the limits of the axis to
change.&nbsp; For example, if you hold the right mouse button and drag
to the right, the limits of the x-axis will change to smaller values,
effectively 'zooming in'.<br><br><span style="font-weight: bold;"><a name="zoom_rectangle"></a><br><img style="width: 35px; height: 34px;" alt="" src="graphApp_manual_files/zoom.png">&nbsp;&nbsp; Zoom Rectangle</span><br>Press
once to set the graph to zoom mode.&nbsp; Press again to get out of
zoom mode.&nbsp; While in zoom mode, you can left click and drag, and a
selection box will appear.&nbsp; When you release the left mouse
button, the graph will zoom into the box that was selected.&nbsp;
Zooming with the right mouse button is also possible, but the function
of it is not completely understood.&nbsp; Usage of this feature is not
recommended.<br><br><span style="font-weight: bold;"><a name="zoom_x"></a><br><img style="width: 37px; height: 38px;" alt="" src="graphApp_manual_files/zoom_x.png">&nbsp;&nbsp; Zoom X-Axis</span><br>Zoom
in by only changing the limits of the x-axis.&nbsp; Similar to Zoom
Rectangle except that the height of the selection box will always be
equal to the height of the graph.&nbsp; Good for when you want to zoom
without changing the limits of the y-axis.<br><br><span style="font-weight: bold;"><a name="zoom_y"></a><br><img style="width: 37px; height: 38px;" alt="" src="graphApp_manual_files/zoom_y.png">&nbsp;&nbsp; Zoom Y-Axis</span><br>Zoom
in by only changing the limits of the y-axis.&nbsp; Similar to Zoom
Rectangle except that the width of the selection box will always be
equal to the width of the graph.&nbsp; Good for when you want to zoom
without changing the limits of the x-axis.<br><br><br><span style="font-weight: bold;"><a name="annotate_point"></a><br><img style="width: 38px; height: 34px;" alt="" src="graphApp_manual_files/ann_point.png">&nbsp;&nbsp; Annotate Point</span><br>Once this button is pressed, clicking on any point in the graph will result in that point being annotated.<br><br><span style="font-weight: bold;"><a name="annotate_threshold"></a><br><img style="width: 38px; height: 36px;" alt="" src="graphApp_manual_files/ann_threshold.png">&nbsp;&nbsp; Annotate Threshold</span><br>Use
this to annotate either all peaks above a certain threshold value, or
all valleys below a certain threshold value.&nbsp; When this button is
clicked, and dialog box is displayed asking for the threshold value,
and whether to annotate peaks or valleys.<br><br><span style="font-weight: bold;"><a name="annotate_peak"></a><br><img style="width: 38px; height: 34px;" alt="" src="graphApp_manual_files/ann_peak.png">&nbsp;&nbsp; Annotate Peak</span><br>Annotate
the highest peak in an area.&nbsp; When this button is clicked, you can
drag and create a selection box.&nbsp; When the mouse button is
released, the highest peak in the selection box is annotated.&nbsp; The
definition of a peak is a point where it is higher than both its
neighbors.&nbsp; If a point is at one of the ends of a chromosome, it
only needs to be higher than its only neighbor. &nbsp;<br><br><span style="font-weight: bold;"><a name="annotate_valley"></a><br><img style="width: 39px; height: 33px;" alt="" src="graphApp_manual_files/ann_valley.png">&nbsp;&nbsp; Annotate Valley</span><br>Annotate
the lowest valley in an area.&nbsp; When this button is clicked, you
can drag and create a selection box.&nbsp; When the mouse button is
released, the lowest valley in the selection box is annotated.&nbsp;
The definition of a valley is a point where it is lower than both its
neighbors.&nbsp; If a point is at one of the ends of a chromosome, it
only needs to be lower than its only neighbor. <br><br><span style="font-weight: bold;"><a name="move_annotation"></a><br><img style="width: 45px; height: 40px;" alt="" src="graphApp_manual_files/ann_move.png">&nbsp;&nbsp; Move Annotation</span><br>Press
once to set the graph to move annotation mode.&nbsp; Press again to get
out of the mode.&nbsp; While in this mode, you can click and drag
annotations on the graph and move then wherever you want to.<br><br><span style="font-weight: bold;"><a name="configure_plot"></a><br><img style="width: 44px; height: 36px;" alt="" src="graphApp_manual_files/configure.png">&nbsp;&nbsp; Configure Plot</span><br>Clicking
on this button will cause a dialog box to appear that will let you
change the amount of space between the graph and the window.&nbsp; Will
cause values to be set to default values at first, which is
annoying.&nbsp; May not be too useful, but there if you need it.<br><br><h2><a name="File_Menu_Bar"></a></h2><h2>File Menu Bar</h2><hr style="width: 100%; height: 2px;">&nbsp;&nbsp; &nbsp;<br><br><span style="font-weight: bold;"><a name="fmb_open_graph_file"></a><br>Open GRX File </span>-
Opens a .grx file.&nbsp; This should not be used when you want to
load data from a text file.&nbsp; Use 'Import Data from File' instead.<br><br><span style="font-weight: bold;"><a name="fmb_save_graph_file"></a><br>Save GRX File </span>-
Save the current state of the graph as a .grrx file.&nbsp; The state
of the graph (chromosome, annotations, viewing limits, etc.) will be
saved to a file and can be opened later on using 'Open GRX File'.<br><br><span style="font-weight: bold;"><a name="fmb_save_as_graph_file"></a><br>Save As GRX File</span> - Same as 'Save GRX File', except you can choose a new filename.<br><br><span style="font-weight: bold;"><a name="fmb_import_data_from_file"></a><br>Import Data from File</span>
- Load data to be graphed from a text file.&nbsp; First a file open
dialog box will appear asking you which file you want to load.&nbsp;
After the file is selected, the file format dialog box will
appear.&nbsp; Here is a screenshot of the dialog (Linux version).<br><br><div style="text-align: center;"><img style="width: 712px; height: 559px;" alt="" src="graphApp_manual_files/fileformat_dialog.png"><br></div><br>The
top of the dialog shows a preview of the text file.&nbsp; The first
five lines of the file are shown to help determine which columns to
choose.&nbsp; Note that comments can be used inside a text file.&nbsp;
Comments are denoted with a '#' at the beginning of the line, and will be ignored when processing the file.<br><br>The
next line asks which column indicates position and which indicates ppl
values.&nbsp; Choose the correct columns from the drop down list.
&nbsp;The default for the position column is the second column. The
default for the PPL column is the third column. The third option
specifies how many lines make up the header of the file. For many
files, there is a header at the top that has information about the
file, like what type of information the columns stand for. The default
for the header is one line.<br><br>The
next
part of the dialog asks whether this file has data for a single
chromosome or multiple chromosomes.&nbsp; The default is to use
multiple chromosomes.&nbsp; If single chromosome is chosen, you must
also specify what number that chromosome is.&nbsp; If multiple
chromosomes are chosen, you have the option of specifying the column
that indicates what number chromosome this line of data is associated
with.&nbsp; If 'None' is chosen for this value, then it is assumed that
the data listed is for chromosome one, then chromosome two, etc.&nbsp;
Data for a chromosome ends and a new chromosome is started when the
position value for the current line is less then the position value
from the previous line.&nbsp; The default is to have the first column
be the column that specifies the chromosome number. If multiple
chromosomes are selected, you
can also specify to import all chromosome found in the file, or only
import specific ones.&nbsp; If only specific chromosomes are chosen,
separate different chromosome numbers with a comma (e.g. '1, 4, 8').<br><br>The
next text box will ask what the name of this line set is. A line set is
defined as all the lines imported from one text file. For example, if a
set of 23 chromosomes are imported from one file, that will be counted
as the first line set. Then later on 23 more chromosomes are imported
from another file (using a separate overlap option, for example), and
those 23 are counted as a second set. The name of the set can be shown
by the legend instead of the name of each individual line (see 'Options
Menu Bar', 'Configure Graph'). Also the name of the set is used to
identify the line set when trying to change the line properties of the
sets (see 'Options Menu Bar', 'Configure Line Sets'). This box can be
left blank.<br><br>The
next two text boxes ask whether to attach a prefix or postfix to the
chromosome numbers.&nbsp; For example, if a postfix of 'a' is
specified, then the chromosome number will be listed as 1a, 2a,
etc.&nbsp; Handy for when you overlap different chromosomes together,
and you want their names to be different.<br><br>The next text box asks
if there are any columns with extra data.&nbsp; Any extra data for that
point will be displayed at the lower left corner next to the position
and ppl value of the mouse position.&nbsp; For example, if each
position has an associated marker to it listed in the file, that column
can be listed.&nbsp; Then when moving the pointer over the graph, the
lower left corner will show the marker data of the closest point to the
current mouse position.<br><br>The last line asks whether to keep
imported lines the same color.&nbsp; The default is to have it
checked.&nbsp; The color button next to it lets you choose what one
color to use for all imported chromosomes.&nbsp; If this is unchecked,
each chromosome being imported will each have a different color, and
the color box has no effect.<br><br><span style="font-weight: bold;">Note</span>:
Importing Data from File when there are already some chromosomes loaded
in the graph will result in the chromosomes being placed at the end of
the graph.&nbsp; If you want to overlap different chromosomes together,
see 'Overlap Menu Bar'.<br><br><span style="font-weight: bold;"><a name="fmb_import_data_from_remote_file"></a><br>Import Data from Remote File</span>
- Load a data file from a url.&nbsp; When this is chosen, a dialog box
will appear asking for the url of the remote file.&nbsp; Simply type
the entire path to the file (including http://), then the same file
format dialog will appear as shown earlier.<br><br><span style="font-weight: bold;"><a name="fmb_export_graph_to_color_image"></a><br>Export Graph to Color Image</span>
- Save the current view as an image file.&nbsp; This is equivalent to
clicking on the Save button on the toolbar.&nbsp; The preferred image
format is png.<br><br><span style="font-weight: bold;"><a name="fmb_export_graph_to_grayscale_iamge"></a><br>Export Graph to Grayscale Image</span> - Same as 'Export Graph to Color Image', except that the final image will be grayscale.<br><br><span style="font-weight: bold;"><a name="fmb_page_setup"></a><br>Page Setup</span>
- Page Setup options for printing.&nbsp; Currently lets you set the
width of the figure and the margin.&nbsp; There is no option to choose
between landscape or portrait orientation.&nbsp; The page orientation
is automatically set using the width of the figure.&nbsp; If the width
is too large to fit using portrait, then landscape is used.<br><br><span style="font-weight: bold;"><a name="fmb_print"></a><br>Print</span> - Prints the current graph.<br><br><h2><a name="Edit_Menu_Bar"></a></h2><h2>Edit Menu Bar</h2><hr style="width: 100%; height: 2px;"><br><br><span style="font-weight: bold;"><a name="emb_clear_graph"></a><br>Clear Graph</span> - Resets the graph and removes everything from it.<br><br><span style="font-weight: bold;"><a name="emb_copy_graph"></a><br>Copy Graph</span> - Put the graph on the clipboard.<br><br><span style="font-weight: bold;"><a name="emb_remove_chr"></a><br>Remove Chromosome</span>
- Remove a chromosome from the graph.&nbsp; A dialog box with all
chromosomes in the graph listed will appear.&nbsp; Simply check the
ones you wish to remove, then click OK.<br><br><h2><a name="Options_Menu_Bar"></a></h2><h2>Options Menu Bar</h2><hr style="width: 100%; height: 2px;"><br><br><span style="font-weight: bold;"><a name="omb_configure_graph"></a><br>Configure Graph</span>
- Lets you configure many aspects of how the graph is drawn.&nbsp; The
first three text boxes allow you to add a title, and labels to the x
and y-axis. Then there are options that affect the entire graph, such
as
background color and the text size of the title and axis labels.&nbsp;
There
are
also options that allow you to show a legend for the graph, which is
off by default.&nbsp; Next to that is an option as to what the legend
show. The legend can display each individual line's name, or it can
show only the name of the different line sets. The next row includes
options about the legend, specifically whether to show a frame around
the legend, and where to place the legend. The next two rows of options
deal with
annotations.
The first row has options to set the precision of the ppl value in the
annotation text. If set to default, two decimal places are shown,
except when the ppl value is less than 0.02, in which case three
decimal places are show. If that is not desired, then a custom
precision can be set. The next line has options to set the precision in
regards to the position text in annotations. The next option lets you
display annotations the
same color as the line associated with it. Normally annotations are all
black. &nbsp;The option on the next line lets you suppress duplicate
line numbers in the labels below the graph. &nbsp;For example, if the
label for a group of lines is normally '1, 1, 1', then choosing this
option makes the label simply '1'. This option is helpful when
overlapping many lines that have the same chromosome &nbsp;number.<br><br>The bottom of the dialog has a list of all chromosome in the
graph, with options to change attributes of a chromosome and the way
they are drawn. See
below for a&nbsp;screenshot of the dialog box (Linux version).<br><br><span style="font-weight: bold;">Note</span>:
for the title and the axis text, you have the ability to use
mathematical expressions in them.&nbsp; See 'Mathematical Expressions'
for more info.<br><br><div style="text-align: center;"><img style="width: 680px; height: 742px;" alt="" src="graphApp_manual_files/configure_dialog.png"><br></div><span style="font-weight: bold;"><br><a name="omb_configure_linesets"></a><br>Configure Line Sets </span>-
Sets properties of&nbsp;lines sets. &nbsp;The options that can be
changed include the set name, color, line style, thickness, symbol, and
symbol size. When the legend is set to show the name of the line sets,
the name shown is the one in the first column. A screenshot of the
dialog box is shown below (Linux version).<br><br><div style="text-align: center;"><img style="width: 640px; height: 244px;" alt="" src="graphApp_manual_files/configure_linesets.png"><br></div><br><span style="font-weight: bold;"><a name="omb_configure_all_lines"></a><br>Configure All Lines </span>-
Sets properties of all lines at once. This is useful if you want to set
the color, line style, or other properties of lines all at once without
having to change each line manually. <br><br><div style="text-align: center;"><img style="width: 522px; height: 172px;" alt="" src="graphApp_manual_files/configure_all_lines.png"><br></div><span style="font-weight: bold;"><br><a name="omb_configure_genemarkers"></a><br>Configure Gene Markers</span>
- Lets you change certain properties of gene markers (See Insert Menu
Bar -&gt; Gene Markers for more information). &nbsp;A screenshot of the
dialog box is below (Linux version). Each row represents one set of
gene markers. The properties availabe to be configured are the gene
marker's name, size, color, and symbol.<br><br><div style="text-align: center;"><img style="width: 440px; height: 204px;" alt="" src="graphApp_manual_files/genemarkers_config.png"><br></div><span style="font-weight: bold;"><br><a name="omb_set_view_limits"></a><br>Set x-axis, y-axis limits</span>
- Lets you exactly determine the viewing limits of the x-axis and
y-axis. A screenshot of the dialog box is below (Linux version). There
are four textboxes, one for xmin, xmax, ymin, and ymax. For xmin and
xmin, the values in the textbox are relative to a chromosome in the
graph. Next to the xmin and xmax textboxes are choice boxes that let
you select which chromosome to use as a reference point. The range
column displays the current range of positions for the currently
selected chromsome. You cannot use an x value that is outside the range
of the chromosome. For y-axis values, simply type in the number you
wish the view limits to be.<br><br><div style="text-align: center;"><img style="width: 378px; height: 223px;" alt="" src="graphApp_manual_files/view_limits_dialog.png"><br></div><h2><a name="Insert_Menu_Bar"></a></h2><h2>Insert Menu Bar</h2><hr style="width: 100%; height: 2px;"><br><span style="font-weight: bold;"><a name="imb_insert_hline"></a><br>Insert Horizontal Line</span>
- Insert a horizontal line that goes through the entire graph.&nbsp;
See the toolbar button 'Insert horizontal line' for more info.<br><br><span style="font-weight: bold;"><a name="imb_insert_genemarkers"></a><br>Insert Gene Markers</span>
- Insert a set of gene markers. A screenshot of the dialog box is shown
below (Linux version). Gene markers are specific points on the
chromosome that are marked by a certain symbol. The default is to use a
circle as the symbol. When this option is chosen, a file dialog box
will appear, asking for the text file to load. The file containing gene
markers must have a least two columns. One column must indicate the
chromosome number, and the other column must indicate the position
along the chromosome of the gene marker. The dialog box also allows you
to give the set of gene markers a name, and change their size, color
and symbol. &nbsp;The gene marker name is only useful in the case when
you load multiple sets of gene markers and later configure them. The
name will be displayed so that you can differentiate between them.<br><br><div style="text-align: center;"><img style="width: 471px; height: 337px;" alt="" src="graphApp_manual_files/genemarkers_insert.png"><br></div><br><h2><a name="Overlap_Menu_Bar"></a></h2><h2>Overlap Menu Bar</h2><hr style="width: 100%; height: 2px;"><br><span style="font-weight: bold;"><a name="omb_insert_overlap_chr_from_file"></a><br>Insert Overlapping Chromosome from File</span>
- Import chromosome(s) from a text file, and overlap these new
chromosomes with existing chromosomes in the graph.&nbsp; When you
first click this option, a file open dialog will appear, and you choose
what file to import the data from.&nbsp; Once the file is chosen, a
file format dialog appears that is almost identical to the one shown
earlier.&nbsp; The only difference is that at the bottom there is a
drop list with the label 'Chromosome to begin overlap with'.&nbsp; Here
you choose which existing chromosome where the overlapping will
begin.&nbsp; For example, if the graph is currently loaded with
chromosome 3, 4, 7, 12 and you import chromosomes 9, 14, 20 from a file
and choose chromosome 4 as the chromosome to begin overlapping with,
then chromosome 9 will overlap with 4, 14 will overlap with 7, and 20
will overlap with 12.<br><br><span style="font-weight: bold;"><a name="omb_overlap_existing_chr"></a><br>Overlap existing chromosomes</span>
- Takes an existing chromosome(s) from the graph and moves it so it
overlaps with another existing chromosome.&nbsp; When you choose this
option, a dialog box like the one below will appear.&nbsp; At the top
is a list of all chromosomes currently in the graph.&nbsp; You can
select as many as you want to move.&nbsp; At the bottom is a drop down
box that lets you select where to move all the previously selected
chromosomes.<br><br><h2><a name="Annotations"></a></h2><h2>Annotations</h2><hr style="width: 100%; height: 2px;"><br><br>Annotations are used to highlight and label points of interest.<br><br><br><span style="font-weight: bold;"><a name="ann_add_annotation"></a><br>Add Annotation</span><br>To add an annotation, use one of the annotation buttons on the toolbar.&nbsp; See 'Toolbar Buttons' for more info.<br><br><br><span style="font-weight: bold;"><a name="ann_remove_annotation"></a><br>Remove Annotation</span><br>To remove an annotation, simply right-click on the annotation, and select Remove Annotation.<br><br><div style="text-align: center;"><img style="width: 345px; height: 255px;" alt="" src="graphApp_manual_files/remove_annotation.png"><br></div><br><span style="font-weight: bold;"><a name="ann_edit_annotations"></a><br>Editing Annotations</span><br>The
default text for an annotation is to display the ppl value as a
percent, then the '@' symbol, then the position of that point in
centimorgans.&nbsp; To edit the text of an annotation, simple
right-click on the annotation and select Edit Annotation.&nbsp; Note
that you can use mathematical expressions for annotation text.&nbsp;
See 'Mathematical Expressions'.<br><br><span style="font-weight: bold;"><a name="ann_annotation_layout"></a><br></span><span style="font-weight: bold;">Annotation Layout</span><br>There
is an automatic layout algorithm for the annotations.&nbsp; It tries to
place annotations in such a way that it will not overlap with other
annotations or lines in the graph.&nbsp; The algorithm is not perfect,
and will occasionally break, especially when there are many annotations
in a neighborhood.&nbsp; The best remedy for fixing this is to manually
move the annotations yourself.&nbsp; (See 'Move Annotation' in Toolbar
Buttons).&nbsp; Once you move an annotation, you are forever
responsible for it. The automatic layout algorithm will not move it
anymore. This applies as long as you are in the same view. You go to a
different view, for example, when you zoom in. If you pan, the
annotations will stay in place.&nbsp; They will also remember positions
between views when using the back and forward buttons to change the
view.&nbsp; The automatic layout algorithm will initiate when a new
view is created.<br><br><h2><a name="command_line"></a></h2><h2>Command Line Options</h2><hr style="width: 100%; height: 2px;"><br><br>There are command line options available to perform most actions in graphApp.<br>The syntax is to type the following at the command line:<br><br>graphApp &lt;command1&gt; &lt;args1&gt; &lt;command2&gt; &lt;args2&gt; etc...<br><br>If a command appears multiple times, only the first occurance is used, unless<br>the command works on only one chr, in which case it can be&nbsp; specified multiple<br>times. Other commands that can be used multiple times are specified in their<br>description.<br><br>Commands are performed in the order that they are specified, except in some<br>special cases:<br>&nbsp;&nbsp;&nbsp; -open is always performed first.<br>&nbsp;&nbsp;&nbsp; -save and -export are always performed last.<br><br>Command arguments that are column indices start at 1, not 0.<br><br>Some command line arguments ask to specify a color. There are three options<br>available to specify color:<br>&nbsp;&nbsp;&nbsp; 1. Some colors are identified by a single letter. All available colors like<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; this are:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'b' - blue, 'g' - green, 'r' - red, 'c' - cyan, 'm' - magenta, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 'y' - yellow, 'k' - black, 'w' - white<br>&nbsp;&nbsp;&nbsp; 2. You can specify colors by a hex value, for example '#0022ff'<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (Note the '#' instead of 'x' for a hex value)<br>&nbsp;&nbsp;&nbsp; 3. You can specify a name of a color in the X11 color set, for example<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 'Indigo', 'PapayaWhip', or 'Olive'. For a listing of what colors are<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; available, click <a href="graphApp_manual_files/x11color.htm" target="_blank">here</a>, or goto <br>&nbsp; &nbsp; &nbsp; &nbsp;<a href="http://www.mcfedries.com/books/cightml/x11color.htm">http://www.mcfedries.com/books/cightml/x11color.htm</a>.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Note that all colors may not work.<br><br><br><span style="font-weight: bold;"></span>-open &lt;type&gt; &lt;filename&gt; <br>&nbsp;&nbsp;&nbsp; * &lt;type&gt; --&gt; 'text', 'graph', 'grx', what type of file to open<br>&nbsp;&nbsp;&nbsp; * &lt;filename&gt; --&gt; path to file<br><br>&nbsp;&nbsp;&nbsp; * all these commands are used when &lt;type&gt; is 'text'<br>&nbsp;&nbsp;&nbsp; * command are optional unless otherwise specified<br>&nbsp;&nbsp;&nbsp; <br>&nbsp;&nbsp;&nbsp; -pos_col &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * column index of column with position values<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the column index<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * required<br><br>&nbsp;&nbsp;&nbsp; -ppl_col &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * column index of column with ppl values<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the column index<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * required<br>&nbsp;&nbsp;&nbsp; <br>-header_lines &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * the number of lines at the top of the file that constitute the header.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * the header will be ignored when processing the file.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the number of lines<br><br>&nbsp;&nbsp;&nbsp; -chr_num &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * if text file has info about only one chr, tells program what the <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; chr number of that chr is<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the number of the chr in the file<br><br>&nbsp;&nbsp;&nbsp; -chr_col &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * column index of column with chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the column index<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * optional, if left out, will assume first set of values is chr1, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; then second set of values is chr2, etc.<br><br>&nbsp;&nbsp;&nbsp; -chr_select &lt;num1&gt; &lt;num2&gt; etc.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * select only some chr's to be loaded in<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num1&gt; --&gt; integer, the number of the chr that should be loaded<br><br>&nbsp;&nbsp;&nbsp; -chr_prefix &lt;prefix&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * changes name of chr to add a prefix string before every chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;prefix&gt; --&gt; string<br><br>&nbsp;&nbsp;&nbsp; -chr_postfix &lt;postfix&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * changes name of chr to add a postfix string after every chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;postfix&gt; --&gt; string<br><br>&nbsp;&nbsp;&nbsp; -extra_data_col &lt;num1&gt; &lt;num2&gt; etc.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * designate columns that contain extra data associated with each point<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * used to specify a marker # for each point, then when moving the<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cursor, the graph would display what marker # the closest point to<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; the mouse curser is at<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num1&gt; --&gt; integer, the column index<br><br>&nbsp;&nbsp;&nbsp; -chr_same_color &lt;bool&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * tells program to either keep all chr's the same color, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; or to make each chr a different color<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;bool&gt; --&gt; 'on', 'off'<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * optional, if not present, value is 'on'<br><br>&nbsp;&nbsp;&nbsp; -chr_color_all &lt;color&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * set the same color for all chrs<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * only works if -chr_same_color is 'on'<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br><br>&nbsp;&nbsp;&nbsp; -lineset_name &lt;name&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * gives a name to this line set<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; string, the name of the set<br><br>-save &lt;filename&gt;<br>&nbsp;&nbsp;&nbsp; * save a graph as a .grx file<br>&nbsp;&nbsp;&nbsp; * &lt;filename&gt; --&gt; path to a file<br>&nbsp;&nbsp;&nbsp; * if .grx is not at end of filename, it is added on automatically<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-save_graph &lt;filename&gt;<br>&nbsp;&nbsp;&nbsp; * save a graph as a .graph file<br>&nbsp;&nbsp;&nbsp; * &lt;filename&gt; --&gt; path to a file<br>&nbsp;&nbsp;&nbsp; * if .graph is not at end of filename, it is added on automatically<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-export &lt;filename&gt;<br>&nbsp;&nbsp;&nbsp; * export graph to an image<br>&nbsp;&nbsp;&nbsp; * &lt;filename&gt; --&gt; path to a file<br>&nbsp;&nbsp;&nbsp; * file format is determined by file extension in filename. only .eps <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; and .png files work. if no file extension, or an unrecognized one, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; .png will be added to filename<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-noshow<br>&nbsp;&nbsp;&nbsp; * does not show the window<br>&nbsp;&nbsp;&nbsp; * closes the program as soon as the other commands are done<br><br>-comment &lt;comment&gt;<br>&nbsp;&nbsp;&nbsp; * specify the comments on the side of the graph<br>&nbsp;&nbsp;&nbsp; * &lt;comment&gt; --&gt; a string (be sure to put quotes around the comment)<br><br>-hline &lt;num1&gt; &lt;num2&gt; etc.<br>&nbsp;&nbsp;&nbsp; * specify horizontal line(s) in graph<br>&nbsp;&nbsp;&nbsp; * &lt;num1&gt; --&gt; number, the y-value (ppl-value) of the line(s)<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br>&nbsp;&nbsp;&nbsp; <br>-title &lt;label&gt;<br>&nbsp;&nbsp;&nbsp; * set the title of the graph, will appear above graph<br>&nbsp;&nbsp;&nbsp; * &lt;label&gt; --&gt; string<br><br>-xlabel &lt;label&gt;<br>&nbsp;&nbsp;&nbsp; * set the label of the x-axis<br>&nbsp;&nbsp;&nbsp; * &lt;label&gt; --&gt; string<br><br>-ylabel &lt;label&gt;<br>&nbsp;&nbsp;&nbsp; * set the label of the y-axis<br>&nbsp;&nbsp;&nbsp; * &lt;label&gt; --&gt; string<br><br>-bgcolor &lt;color&gt;<br>&nbsp;&nbsp;&nbsp; * set background color<br>&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br><br>-title_text_size &lt;num&gt;<br>&nbsp;&nbsp;&nbsp; * the font size of the graph title<br>&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer<br>&nbsp;&nbsp;&nbsp; * default is 14<br><br>-axis_text_size &lt;num&gt;<br>&nbsp;&nbsp;&nbsp; * the font size of the x and y-axis text<br>&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer<br>&nbsp;&nbsp;&nbsp; * default is 12<br><br>-show_legend <br>&nbsp;&nbsp;&nbsp; * show the legend <br>&nbsp;&nbsp;&nbsp; * default is to not show legend<br><br>-show_legend_frame &lt;bool&gt;<br>&nbsp;&nbsp;&nbsp; * show a frame around the legend<br>&nbsp;&nbsp;&nbsp; * &lt;bool&gt; --&gt; 'on' or 'off'<br>&nbsp;&nbsp;&nbsp; * default is 'on'<br><br>-legend_show_linesets &lt;bool&gt;<br>&nbsp;&nbsp;&nbsp; * &lt;bool&gt; --&gt; 'on' or 'off'<br>&nbsp;&nbsp;&nbsp; * if 'on', the legend lists the line sets<br>&nbsp;&nbsp;&nbsp; * if 'off', the legend lists each individual line<br>&nbsp;&nbsp;&nbsp; * default is 'off'<br><br>-legend_location &lt;loc&gt;<br>&nbsp;&nbsp;&nbsp; * the location of the legend<br>&nbsp;&nbsp;&nbsp; * &lt;loc&gt; --&gt; 'upper right' , 'upper left', 'upper center', 'lower right',&nbsp; <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'lower left', 'lower center', 'center left',&nbsp; 'center right', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 'right', 'center'<br>&nbsp;&nbsp;&nbsp; * default is 'upper right'<br><br>-ann_ppl_precision &lt;arg1&gt;<br>&nbsp;&nbsp;&nbsp; * specifies the amount of ppl precision to display in annotations<br>&nbsp;&nbsp;&nbsp; * &lt;arg1&gt; --&gt; 'default' or integer<br>&nbsp;&nbsp;&nbsp; * with 'default', ppl is shown with two decimal places (e.g. 7%), except<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; when the ppl is less than .02, in which can it is shown to three <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; decimal places (e.g. 7.2%)<br><br>-ann_pos_precision &lt;num&gt;<br>&nbsp;&nbsp;&nbsp; * specifies how many decimal places to show for the position number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; in the text of an annotation<br>&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, how many decimal places to show<br>&nbsp;&nbsp;&nbsp; * default is 0<br><br>-ann_same_color<br>&nbsp;&nbsp;&nbsp; * specifies that annotations should be the same color as the line it is<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; associated with. only works with annotations specifying peaks and valleys<br>&nbsp;&nbsp;&nbsp; * default is that all annotations are black<br><br>-no_duplicate_numbers<br>&nbsp;&nbsp;&nbsp; * when chromosome numbers overlap and they are the same number,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; do not show the number multiple times in the labels underneath<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; the graph. for example, '1, 1, 1' becomes '1'<br><br>-lineset_change_name &lt;oldname&gt; &lt;newname&gt;<br>&nbsp;&nbsp;&nbsp; * change the name of a line set, changes the name in the legend if the<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; legend is set to display line set names <br>&nbsp;&nbsp;&nbsp; * &lt;oldname&gt; --&gt; the old name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;newname&gt; --&gt; the new name of the line set<br><br>-lineset_color &lt;name&gt; &lt;color&gt;<br>&nbsp;&nbsp;&nbsp; * change the color of all chrs in a line set<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br><br>-lineset_linestyle &lt;name&gt; &lt;style&gt;<br>&nbsp;&nbsp;&nbsp; * set the line style of all chrs in a line set<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;style&gt; --&gt; the style of the line<br>&nbsp;&nbsp;&nbsp; * valid styles are: 'solid', 'dashed', 'dotted', 'dash-dot', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'steps', 'no line'<br>&nbsp;&nbsp;&nbsp; * default is 'solid'<br><br>-lineset_thickness &lt;name&gt; &lt;thickness&gt;<br>&nbsp;&nbsp;&nbsp; * change the thickness of all chrs in a line set<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;thickness&gt; --&gt; integer, how thick the chr lines should be<br>&nbsp;&nbsp;&nbsp; * default is 2<br><br>-lineset_symbol &lt;name&gt; &lt;symbol&gt;<br>&nbsp;&nbsp;&nbsp; * change the symbol that represents each data point<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;symbol&gt; --&gt; symbol<br>&nbsp;&nbsp;&nbsp; * valid symbols are: 'no symbol', 'square', '+', 'o', 'x', 'diamond', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'thin diamond', '^', 'v', '&gt;', '&lt;', '|', '_', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'hexagon', 'pentagon', 'tripod 1', 'tripod 2'<br>&nbsp;&nbsp;&nbsp; * default is 'no symbol'<br><br>-lineset_symbol_size &lt;name&gt; &lt;size&gt;<br>&nbsp;&nbsp;&nbsp; * change the size of the symbol for all chrs in a line set<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the line set<br>&nbsp;&nbsp;&nbsp; * &lt;size&gt; --&gt; integer, new size of the symbol<br>&nbsp;&nbsp;&nbsp; * default is 6<br><br>-chr_name &lt;oldname&gt; &lt;newname&gt;<br>&nbsp;&nbsp;&nbsp; * change the name of a chr, changes what is displayed at the <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; bottom of the graph<br>&nbsp;&nbsp;&nbsp; * &lt;oldname&gt; --&gt; the old name of the chr, when chr's are first loaded,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
their name is simply their chr number, unless a prefix or<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
postfix was applied<br>&nbsp;&nbsp;&nbsp; * &lt;newname&gt; --&gt; the new name of the chr<br><br>-chr_color &lt;name&gt; &lt;color&gt;<br>&nbsp;&nbsp;&nbsp; * change the color of a chr<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the chr<br>&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br><br>-chr_linestyle &lt;name&gt; &lt;style&gt;<br>&nbsp;&nbsp;&nbsp; * set the line style of a chr<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the chr<br>&nbsp;&nbsp;&nbsp; * &lt;style&gt; --&gt; the style of the line<br>&nbsp;&nbsp;&nbsp; * valid styles are: 'solid', 'dashed', 'dotted', 'dash-dot', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'steps', 'no line'<br>&nbsp;&nbsp;&nbsp; * default is 'solid'<br><br>-chr_thickness &lt;name&gt; &lt;thickness&gt;<br>&nbsp;&nbsp;&nbsp; * change the thickness of a chr<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the chr<br>&nbsp;&nbsp;&nbsp; * &lt;thickness&gt; --&gt; integer, how thick the chr line should be<br>&nbsp;&nbsp;&nbsp; * default is 2<br><br>-chr_symbol &lt;name&gt; &lt;symbol&gt;<br>&nbsp;&nbsp;&nbsp; * change the symbol that represents each data point<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the chr<br>&nbsp;&nbsp;&nbsp; * &lt;symbol&gt; --&gt; symbol<br>&nbsp;&nbsp;&nbsp; * valid symbols are: 'no symbol', 'square', '+', 'o', 'x', 'diamond', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'thin diamond', '^', 'v', '&gt;', '&lt;', '|', '_', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'hexagon', 'pentagon', 'tripod 1', 'tripod 2'<br>&nbsp;&nbsp;&nbsp; * default is 'no symbol'<br><br>-chr_symbol_size &lt;name&gt; &lt;size&gt;<br>&nbsp;&nbsp;&nbsp; * change the size of the symbol for a chr<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the chr<br>&nbsp;&nbsp;&nbsp; * &lt;size&gt; --&gt; integer, new size of the symbol<br>&nbsp;&nbsp;&nbsp; * default is 6<br><br>-view_limits &lt;chr name&gt; &lt;pos&gt; &lt;chr name2&gt; &lt;pos2&gt; &lt;ymin&gt; &lt;ymax&gt;<br>&nbsp;&nbsp;&nbsp; * sets the viewing limit of the graph. the first two arguments specify<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; xmin, the next two specify xmax, and the last two specify the limits<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; for the y-axis<br>&nbsp;&nbsp;&nbsp; * &lt;chr name&gt;, &lt;chr name2&gt; --&gt; label, name of chr's<br>&nbsp;&nbsp;&nbsp; * &lt;pos&gt;, &lt;pos2&gt; --&gt; number, the position along the specified chromosome<br>&nbsp;&nbsp;&nbsp; * &lt;ymin&gt; --&gt; number, the minimum value of the y-axis<br>&nbsp;&nbsp;&nbsp; * &lt;ymax&gt; --&gt; number, the maximum value of the y-axis<br><br>-xlimits &lt;chr name&gt; &lt;pos&gt; &lt;chr name&gt; &lt;pos&gt;<br>&nbsp;&nbsp;&nbsp; * set the x limits of the graph view, this takes two sets of chr names<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; and a position along that chr for the limits. the first set specifies<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; xmin, and the second set specifies xmax<br>&nbsp;&nbsp;&nbsp; * &lt;chr name&gt; --&gt; label, the name of a chr<br>&nbsp;&nbsp;&nbsp; * &lt;pos&gt; --&gt; number, a position along that chr<br><br>-ylimits &lt;ymin&gt; &lt;ymax&gt;<br>&nbsp;&nbsp;&nbsp; * set the y limits of the graph view<br>&nbsp;&nbsp;&nbsp; * &lt;ymin&gt; --&gt; number, the minimum y-value<br>&nbsp;&nbsp;&nbsp; * &lt;ymax&gt; --&gt; number, the maximum y-value<br><br>-zoom_chr &lt;chr name&gt;<br>&nbsp;&nbsp;&nbsp; * set x-limits such that they start and end at a particular chr<br>&nbsp;&nbsp;&nbsp; * &lt;chr name&gt; --&gt; label, the name of the chr to zoom into<br>&nbsp;&nbsp;&nbsp; * Note that this does not affect the y-limits of the graph<br><br>-gene_markers &lt;file&gt; &lt;chr col&gt; &lt;pos col&gt; &lt;size&gt; &lt;color&gt; &lt;symbol&gt; &lt;name&gt;<br>&nbsp;&nbsp;&nbsp; * insert gene markers that are specified by a file<br>&nbsp;&nbsp;&nbsp; * &lt;file&gt; --&gt; path to file, the file that the gene marker data is in<br>&nbsp;&nbsp;&nbsp; * &lt;chr col&gt; --&gt; integer, the column index that specifies what chr this <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; gene marker belongs to<br>&nbsp;&nbsp;&nbsp; * &lt;pos col&gt; --&gt; integer, the column that indicates the position of the gene<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; marker<br>&nbsp;&nbsp;&nbsp; * &lt;size&gt; --&gt; integer, the size of the markers, default is 4<br>&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color, the color of the markers, default is 'r' (red)<br>&nbsp;&nbsp;&nbsp; * &lt;symbol&gt; --&gt; symbol, the symbol used for the markers, default is 'o'<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, a name given to this set of gene markers<br>&nbsp;&nbsp;&nbsp; * every argument after &lt;pos col&gt; is optional, but needs to be in the<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; correct order and cannot have any missing parameters. for example, you<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; could specify &lt;size&gt; and &lt;color&gt;, but not &lt;size&gt; and &lt;symbol&gt; by itself,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; you would need to specify &lt;size&gt; &lt;color&gt; &lt;symbol&gt;<br>&nbsp;&nbsp;&nbsp; * valid symbols are: 'no symbol', 'square', '+', 'o', 'x', 'diamond', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'thin diamond', '^', 'v', '&gt;', '&lt;', '|', '_', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'hexagon', 'pentagon', 'tripod 1', 'tripod 2'<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-gene_markers_size &lt;name&gt; &lt;size&gt;<br>&nbsp;&nbsp;&nbsp; * changes the size of a set of gene markers<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the set of gene markers to change<br>&nbsp;&nbsp;&nbsp; * &lt;size&gt; --&gt; integer, the new size of the gene makers<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-gene_markers_color &lt;name&gt; &lt;color&gt;<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the set of gene markers to change<br>&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-gene_markers_symbol &lt;name&gt; &lt;symbol&gt;<br>&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; label, the name of the set of gene markers to change<br>&nbsp;&nbsp;&nbsp; * &lt;symbol&gt; --&gt; symbol, the new symbol of the gene markers<br>&nbsp;&nbsp;&nbsp; * valid symbols are: 'no symbol', 'square', '+', 'o', 'x', 'diamond', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'thin diamond', '^', 'v', '&gt;', '&lt;', '|', '_', <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
'hexagon', 'pentagon', 'tripod 1', 'tripod 2'<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-overlap &lt;filename&gt; &lt;pos col&gt; &lt;ppl col&gt; &lt;overlap chr&gt;<br>&nbsp;&nbsp;&nbsp; * reads chromosome information from a file, and overlaps them with<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; other chromosomes in the graph<br>&nbsp;&nbsp;&nbsp; * &lt;filename&gt; --&gt; path to file, the file with the chr data, must be text file<br>&nbsp;&nbsp;&nbsp; * &lt;pos col&gt; --&gt; integer, the column index of column with position values<br>&nbsp;&nbsp;&nbsp; * &lt;ppl col&gt; --&gt; integer, the column index of column with ppl values<br>&nbsp;&nbsp;&nbsp; * &lt;overlap chr&gt; --&gt; label, the name of a current chr in the graph that<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
the overlap will begin at<br>&nbsp;&nbsp;&nbsp; * &lt;overlap chr&gt; is optional, if not present, the overlap will begin at<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; the first (leftmost) chr<br><br>&nbsp;&nbsp;&nbsp; * all these commands are used to specifiy extra information <br>&nbsp;&nbsp;&nbsp; * about the file being opened by -overlap. they should appear<br>&nbsp;&nbsp;&nbsp; * right after -overlap to ensure that they are assoicated with<br>&nbsp;&nbsp;&nbsp; * that particular -overlap (just in case there are multiple -overlaps)<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br>&nbsp;&nbsp;&nbsp; <br>&nbsp;&nbsp;&nbsp; -overlap_chr_num &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * if text file has info about only one chr, tells program what the <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; chr number of that chr is<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num&gt; --&gt; integer, the number of the chr in the file<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_col &lt;num&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * column index of column with chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * optional, if left out, will assume first set of values is chr1, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; then second set of values is chr2, etc.<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_select &lt;num1&gt; &lt;num2&gt; etc.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * select only some chr's to be loaded in<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num1&gt; --&gt; integer, the number of the chr that should be loaded<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_prefix &lt;prefix&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * add a prefix string before every chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;prefix&gt; --&gt; string<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_postfix &lt;postfix&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * add a postfix string after every chr number<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;postfix&gt; --&gt; string<br><br>&nbsp;&nbsp;&nbsp; -overlap_extra_data_col &lt;num1&gt; &lt;num2&gt; etc.<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * designate columns that contain extra data associated with each point<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * used to specify a marker # for each point, then when moving the<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cursor, the graph would display what marker # the closest point to<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; the mouse curser is at<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;num1&gt; --&gt; integer, the column index<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_same_color &lt;bool&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * tells program to either keep all chr's the same color, <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; or to make each chr a different color<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;bool&gt; --&gt; 'on', 'off'<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * optional, if not present, value is 'on'<br><br>&nbsp;&nbsp;&nbsp; -overlap_chr_color_all &lt;color&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * set the same color for all chrs<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;color&gt; --&gt; color<br><br>&nbsp;&nbsp;&nbsp; -overlaap_lineset_name &lt;name&gt;<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * gives a name to this line set<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; * &lt;name&gt; --&gt; string, the name of the set<br><br>-ann_threshold &lt;type&gt; &lt;threshold&gt;<br>&nbsp;&nbsp;&nbsp; * create annotations for all peaks or valleys above or below a value<br>&nbsp;&nbsp;&nbsp; * &lt;type&gt; --&gt; 'peak' or 'valley'<br>&nbsp;&nbsp;&nbsp; * &lt;threshold&gt; --&gt; number, the threshold value. if &lt;type&gt; is 'peak',<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; all peaks above threshold are annotated, if &lt;type&gt; is 'valley',<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; all valleys below threshold are annotated<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-ann_peak &lt;name1&gt; &lt;pos1&gt; &lt;name2&gt; &lt;pos2&gt;<br>&nbsp;&nbsp;&nbsp; * annotate the highest peak in a given region, the region is defined<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; by a starting point and an ending point along the x-axis<br>&nbsp;&nbsp;&nbsp; * &lt;name1&gt; --&gt; the name of the chr that has the starting point<br>&nbsp;&nbsp;&nbsp; * &lt;pos1&gt; --&gt; number, a position relative to the &lt;name1&gt; chr<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; defines the exact starting point<br>&nbsp;&nbsp;&nbsp; * &lt;name2&gt; --&gt; the name of the chr that has the ending point<br>&nbsp;&nbsp;&nbsp; * &lt;pos2&gt; --&gt; number, a position relative to the &lt;name2&gt; chr<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; defines the exact ending point<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-ann_valley &lt;name1&gt; &lt;pos1&gt; &lt;name2&gt; &lt;pos2&gt;<br>&nbsp;&nbsp;&nbsp; * annotate the lowest valley in a given region, the region is defined<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; by a starting point and an ending point along the x-axis<br>&nbsp;&nbsp;&nbsp; * &lt;name1&gt; --&gt; the name of the chr that has the starting point<br>&nbsp;&nbsp;&nbsp; * &lt;pos1&gt; --&gt; number, a position relative to the &lt;name1&gt; chr<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; defines the exact starting point<br>&nbsp;&nbsp;&nbsp; * &lt;name2&gt; --&gt; the name of the chr that has the ending point<br>&nbsp;&nbsp;&nbsp; * &lt;pos2&gt; --&gt; number, a position relative to the &lt;name2&gt; chr<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; defines the exact ending point<br>&nbsp;&nbsp;&nbsp; * can be specified multiple times<br><br>-graph2grx &lt;dir/file&gt; &lt;recursive&gt;<br>&nbsp;&nbsp;&nbsp; * converts a .graph to a .grx file. makes a new file with the same name,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; except the file extension changes from .graph to .grx. does not delete<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; the .graph file.<br>&nbsp;&nbsp;&nbsp; * &lt;dir/file&gt; --&gt; either a filename or a directory. if a filename, will<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; convert the .graph file. if its a directory, then will convert all .graph<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; files in that directory<br>&nbsp;&nbsp;&nbsp; * &lt;recursive&gt; --&gt; 'on' or 'off', if 'on', will perform the conversion<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; recursively in all subdirectories<br>&nbsp;&nbsp;&nbsp; * &lt;recursive&gt; is optional, and is off by default. this will only work<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; when a directory is specified for &lt;dir/file&gt;<br><h2><a name="Mathematical_Expressions"></a></h2><h2>Mathematical Expressions</h2><hr style="width: 100%; height: 2px;"><span style="font-weight: bold;"><br><br></span>The
text for the graph and the x and y axis, as well as text for
annotations have the ability to display mathematical symbols and
expressions.&nbsp; The way to create these is to use TeX markup
expressions.&nbsp; If you are going to use math text, you must begin
and end the string with a dollar sign.&nbsp; For example, 'alpha &gt;
beta' will produce plain text, while '$\alpha &gt; \beta$' will produce
the greek letters, as shown below.<br><br><img style="width: 97px; height: 27px;" alt="" src="graphApp_manual_files/math1.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 48px; height: 27px;" alt="" src="graphApp_manual_files/mathtext1.png"><br><br><br>You can use several other TeX symbols, such as <span style="font-family: monospace;">\infty</span>, <span style="font-family: monospace;">\leftarrow</span>, and <span style="font-family: monospace;">\sum</span>.&nbsp; A complete list is shown at the table below.<br><br><img style="width: 150px; height: 23px;" alt="" src="graphApp_manual_files/math2.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 77px; height: 30px;" alt="" src="graphApp_manual_files/mathtext2.png"><br><br><br>To make subscripts and superscripts, use the '_' and '^' symbols, respectively. &nbsp;<br><br><img style="width: 120px; height: 22px;" alt="" src="graphApp_manual_files/math3.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 51px; height: 29px;" alt="" src="graphApp_manual_files/mathtext3.png"><br><br><br>For
some symbols, using a subscript means the text is placed underneath the
symbol, and the superscript is placed above the symbol.&nbsp; This
applies to symbols for which that behavior makes sense.&nbsp; Note how
braces are used for subscripts or superscripts that are longer than one
character.<br><br><img style="width: 139px; height: 25px;" alt="" src="graphApp_manual_files/math4.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 55px; height: 61px;" alt="" src="graphApp_manual_files/mathtext4.png"><br><br><br>The
default font is italics.&nbsp; To use a different font, enclose the
text you want in a different font in a font command, which essentially
means type the font symbol, then the text you want in that font in
braces.&nbsp; For example, to write the word '<span style="font-family: monospace;">sin</span>' in a roman font:<br><br><img style="width: 121px; height: 24px;" alt="" src="graphApp_manual_files/math5.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 57px; height: 32px;" alt="" src="graphApp_manual_files/mathtext5.png"><br><br>Other fonts include roman <span style="font-family: monospace;">\rm</span>, italics <span style="font-family: monospace;">\it</span>, calligraphy <span style="font-family: monospace;">\cal</span>, and typewriter<span style="font-family: monospace;"> \tt</span><br style="font-family: monospace;"><br><img style="width: 342px; height: 24px;" alt="" src="graphApp_manual_files/math6.png">&nbsp;&nbsp; becomes&nbsp; &nbsp;<img style="width: 356px; height: 28px;" alt="" src="graphApp_manual_files/mathtext6.png"><br><br><br>For a more complete reference to what mathematical expressions are available, check out these two links:<br><br><a href="http://matplotlib.sourceforge.net/tutorial.html#mathtext">http://matplotlib.sourceforge.net/tutorial.html#mathtext</a><br><a href="http://matplotlib.sourceforge.net/matplotlib.mathtext.html">http://matplotlib.sourceforge.net/matplotlib.mathtext.html</a><br></body></html>